---
title: "ifCAF_32s_apCAF_36s"
author: "Joshua"
date: '2022-10-31'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Synopsis
DE analysis of 32 Samples from DMXAA ifCAF induction eperiment and 36 samples from IFNg apCAF induction experiment.  Data from Nextseq sequencing run Oct2022. The count data file `ifCAF_32s_apCAF_36s_HT_counts_long.txt` and the metadata file `ifCAF_32s_apCAF_36s_sample_metadata.txt` are available on ArrayExpress with accession number E-MTAB-14943.  

# Dependencies
```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggrepel)
options(dplyr.summarise.inform=FALSE) # Silences an annoying warning by the summarise function
```
Note: DESeq2 call an array of other packages, some of which interfere with the packages of the tidyverse suite. The DEseq2 library is loaded further below

# Data
Raw read count from htSeq is loaded
```{r}
RRC_long<-read.table("ifCAF_32s_apCAF_36s_HT_counts_long.txt", header = T)
```

The gene metadata table contains information about each protein coding gene. The code to generate it is avilable in R markdown `Musmu_GRCm39_Ens109_26Feb2024.Rmd`.
```{r}
Gene_md<-read.table("Musmu_GRCm39.104_protcod_fd_Gene_md.txt", header = F) %>% 
  set_names(nm = c("EnsID", "Gene_Name", "tot_exon_length", "Exon_number", "Chr", "Start", "End", "Strand", "Version", "Source"))
```

The Sample metadata was made manually. It contains information about batches, replicates and biological variables. 
```{r}
Sample_md<-read.table("ifCAF_32s_apCAF_36s_sample_metadata.txt", header = T, sep = "\t")
```


# ifCAF experiment
The samples from the ifCAF induction experiment (s1-s32) are selected
```{r}
Sample_md_ifCAFs<-Sample_md %>% 
  filter(Sample %in% paste("s", 1:32, sep=""))
```

```{r}
RRC_long_ifCAFs <- RRC_long %>%
  filter(Sample %in% Sample_md_ifCAFs$Sample)
```

A wide table of raw read count table is compiled with same gene order and sample order as the metadata. The Sample name is converted to a factor to remain in the same order instead of beingg sorted by alphabetical order. 
```{r}
RRC_ifCAFs<-RRC_long_ifCAFs %>% 
  mutate(Sample=factor(Sample, levels = unique(RRC_long$Sample))) %>% 
  spread(key = Sample, value = RRC) 

RRC_ifCAFs<-Gene_md %>% 
  select(Gene=Gene_Name) %>% 
  left_join(RRC_ifCAFs, by="Gene") %>% 
  select(-EnsID)

RRC_mx_ifCAFs <- RRC_ifCAFs %>% 
  column_to_rownames("Gene") 
```


# DESeq2 differential expression analysis
Here We want to assess the effect of the treatment (i.e the drug) on the transcriptome. We want to do that separately on the tumour cells in monoculture, on the tumour cells from co-culture and on the PSC from co-culture. 

One simple solution would be to subset the data into four separate sets (Mono_T, Co_T Mono_PSC and Co_PSC) and perform the DE analysis spearately. However, this is not recommended because including all the data results in more robust normalization and better model fitting for statistical testing. 

Instead, we will use the simple trick of generating a new variable that incorporates the drug treatment, culture type and cell type. Then, only the comparisons between drug treatments within the same cell type will be done.

If the technical replicates bring very little variance compared to the biological replicates, they should be collapsed. DESeq2 provides a function for that. Here, we will treat the samples from different runs but from the same line as different replicates.  

let's update the sample metadata to include the combined variable. We keep only the info that is useful for DE analysis.  
```{r}
Sample_md_Compvar_ifCAFs <- Sample_md_ifCAFs %>% 
  mutate(Condition=paste(Drug, Culture, Cell_type, sep="_")) %>% 
  select(Sample, Run, Biological.Replicate, Drug, Culture, Cell_type, Condition)
```

The samples are paired in the sense that cells of the same kind tested during the same run vary only by the drug treatment (everything else is equal). 

It is now time to generate the DESeq object and apply the experimental design. Note that DESeq takes the raw read count, as this allows better model fitting for statistical testing than normalized data. Still provides normalized data for visualization purposes. We call it DE-seq Normalized Read Count (DNRC) 


```{r}
library(DESeq2)

dds_ifCAFs <- DESeqDataSetFromMatrix(countData = RRC_mx_ifCAFs,
                              colData = Sample_md_Compvar_ifCAFs,
                              design= ~ Run + Condition)

dds_ifCAFs <- DESeq(dds_ifCAFs)
```

The DEseq-normalized read count (DNRC) is extracted
```{r}
DNRC_ifCAFs <- counts(dds_ifCAFs, normalized=TRUE) %>% 
  as.data.frame() %>% 
  rownames_to_column("Gene")

DNRC_long_ifCAFs <- DNRC_ifCAFs %>% 
  gather(key = Sample, value = DNRC, -Gene)
```

# extraction of q-values
DEseq provides p-values for comparisons between groups of sample as per the applied design. 
It also makes a correction for multiple testing, since there are very many genes. The p-values are adjusted using the False Discovery Rate (FDR) correction. The adjusted p values are called q-values. 

A function is made to extract q values between groups
```{r}
Qs <- function(dds, grouping_var, treat_group, ctrl_group) {
  comp_df <- results(dds, c(grouping_var, treat_group, ctrl_group)) %>%
    as.data.frame() %>%
    rownames_to_column("Gene") %>%
    mutate(FC=2^log2FoldChange) %>% 
    select(Gene, FC, padj) %>%
    mutate(ctrl_group = ctrl_group, treat_group = treat_group)
  
  return(comp_df)
}
```

The q values for Drug-treated vs control comparisons are calculated separately for monocultured tumour cells, co-culture tumour cells monocultured PSC and co-cultured PSC 
DMSO is the control, so placed second in the results function arguments. 
```{r}
FC_qvals_ifCAFs <- Qs(dds_ifCAFs, "Condition", "DMXAA_Co_PSC", "DMSO_Co_PSC") %>% 
  bind_rows(Qs(dds_ifCAFs, "Condition", "DMXAA_Mono_PSC", "DMSO_Mono_PSC")) %>% 
  bind_rows(Qs(dds_ifCAFs, "Condition", "DMXAA_Mono_Tumour", "DMSO_Mono_Tumour")) %>% 
  bind_rows(Qs(dds_ifCAFs, "Condition", "DMXAA_Co_Tumour", "DMSO_Co_Tumour")) %>% 
  bind_rows(Qs(dds_ifCAFs, "Condition", "DMXAA_Co_Tumour", "DMXAA_Mono_Tumour")) %>% 
  bind_rows(Qs(dds_ifCAFs, "Condition", "DMXAA_Co_PSC", "DMXAA_Mono_PSC")) %>% 
  bind_rows(Qs(dds_ifCAFs, "Condition", "DMSO_Co_Tumour", "DMSO_Mono_Tumour")) %>% 
  bind_rows(Qs(dds_ifCAFs, "Condition", "DMSO_Co_PSC", "DMSO_Mono_PSC"))
```

# Selection of DE genes
A cutoff on the absolute expression levels is desirable to remove significant genes that have so low expression values that they are in the range of background noise. 

The mean DNRC per group is introduced
```{r}
group_DNRC_ifCAFs <-DNRC_long_ifCAFs %>% 
  left_join(Sample_md_Compvar_ifCAFs, by = "Sample") %>% 
  group_by(Gene, Condition) %>% 
  summarise(mean_DNRC=round(mean(DNRC), 3)) 

Means_FC_qvals_ifCAFs <- FC_qvals_ifCAFs %>% 
  left_join(group_DNRC_ifCAFs, by=c("Gene","ctrl_group" = "Condition")) %>% 
  dplyr::rename(ctrl_mean_DNRC=mean_DNRC) %>% 
  left_join(group_DNRC_ifCAFs, by=c("Gene","treat_group" = "Condition")) %>% 
  dplyr::rename(treat_mean_DNRC=mean_DNRC) %>% 
  mutate(FC=round(FC, 2)) %>% 
  drop_na()
```

# Drug-modulated genes per condition
The DE genes are extracted with lenient cutoff for plotting purposes
```{r}
LenSel_Means_FC_qvals_ifCAFs<-Means_FC_qvals_ifCAFs %>% 
  filter(map2(ctrl_mean_DNRC, treat_mean_DNRC, max)>3) %>% 
  mutate(FC = FC+0.01) %>%
  mutate(logFC=log(FC)) %>% 
  filter(abs(logFC)>0.1) %>% 
  filter(padj<0.8) %>% 
  arrange(desc(FC))
```

A function to export Z scores
```{r}
export_Z <- function(in_sign, in_Culture, in_Cell_type, in_Drug_list, DNRC_long, sample_list) {
  
  sig_df <- DNRC_long %>%
    mutate_if(is.character,as.factor) %>% 
    filter(Gene %in% in_sign) %>%
    left_join(sample_list, by = "Sample") %>% 
    filter(Culture %in% in_Culture) %>%
    filter(Cell_type %in% in_Cell_type) %>%
    filter(Drug %in% in_Drug_list) %>%
    group_by(Gene) %>%
    mutate(Z = (DNRC - mean(DNRC)) / sd(DNRC)) %>%
    drop_na() %>%
    group_by(Sample) %>%
    summarise(sign_Z = mean(Z) %>% round(2)) %>%
    left_join(sample_list, by = "Sample")
  
 return(sig_df)
}

```

CAF subtype signatures are recovered from the analysis of day6 coculture single cell RNAseq (Figure 3). The code to generate these marker lists is available in R markdown `Figure_3_S6_data_analysis.Rmd`. The lists of markers are also available in supplementary table S6 of the paper. Here, the lists are explicitly loaded directly. 
```{r}
# Cluster 1 from PSC.singlets.rPCA.integrated
myCAF_Sig <- c("Col5a3", "Adcy1", "Pmepa1", "Jag1", "Col1a1", "AA467197", "Itgb1", "Msn", "Klhdc8a", "Npr3", "Cdh11", "Fbn1", "Hbegf", "Slc20a1", "Tln2", "Inhba", "Tfpi2", "Tmsb4x", "Ttll5", "Cnn3", "Runx1", "Sparc", "Myl12a", "Acsbg1", "Acta2", "Csrp2", "Pdgfc", "Ccn4", "Gng11", "Bok", "Ncam1", "Timp3", "Has2", "Tes", "Col8a1", "Plau", "Megf10", "Tmeff1", "Sdc2", "Prkg2", "Serpinh1", "Tgfb3", "Pmaip1", "Ctsc", "Bhlhe41", "Tubb2b", "Creb3l1", "Zfp469", "Vim", "Nrep", "Vegfc", "Tagln", "Atp6v1g1", "Flna", "Postn", "Unc5b", "Col5a1", "Gm6169", "Cav1", "Stk17b", "Rnf149", "Tmsb10", "Plat", "Tars", "Col18a1", "Prss23", "Actg1", "Bhlhe40", "Col5a2", "Palld", "Creb3l2", "Frrs1", "Col6a3", "Crlf1", "Kdelr3", "Wnt5a", "Nrp1", "Uchl1", "Pdgfa", "Anxa2", "Prdx6", "Lrrc59", "Myh9", "Tuba1a", "Epha3", "Tubb3", "Plaur", "Pxdn", "Ccn2", "Crip1", "Ppic", "Bmp1", "Timp1", "Ndfip1", "Basp1", "Cyp51", "Ank", "Fgf2", "Tubb2a", "Cryab", "Vegfa", "Col1a2")

# Cluster 0 from PSC.singlets.rPCA.integrated
iCAF_Sig <- c("Saa3", "Mt2", "Sod2", "Hp", "Mt1", "Cxcl1", "Ptgs2", "Ackr3", "Slc16a1", "Gda", "Ifitm3", "Nfkbiz", "Cxcl5", "Dcn", "Cebpd", "Hif1a", "Sdc4", "Ereg", "Ifitm1", "Phlda1", "Ier3", "C1ra", "Nfkbia", "Cfb", "Gas1", "Cebpb", "Cp", "Dhrs3", "Fgf7", "Serpina3n", "Fth1", "Pi15", "Osmr", "Junb", "Serpina3g", "Fam162a", "C3", "Steap4", "Jak2", "Il4ra", "Nek6", "Nsg1", "Cd74", "Rarres2", "Il13ra1", "Sod3", "Pdpn", "Lcn2", "Dynlrb1", "Mgst1", "Slc9a3r1", "Zfp36l1", "Lgi2", "Slc39a14", "Gm2115", "Zfp281", "Mark1", "Ifi207", "Fos", "Mmp13", "Ch25h", "Ifi27l2a", "Mmp19", "Mmp3", "Fosb", "Cyp1b1", "Serping1", "Nr4a1", "Igfbp3")

# Cluster 4 from PSC.singlets.rPCA.integrated
ifCAF_Sig <- c("Rsad2", "Oasl1", "Cxcl10", "Cmpk2", "Ifit3b", "Ifit3", "Igtp", "Gbp2", "Gbp3", "Ccl5", "Mx1", "Oasl2", "Ifih1", "Ifi44", "Stat2", "Rnd1", "Isg15", "Ifit1", "Phf11b", "Mx2", "Oas1b", "Tgtp1", "Usp18", "Gbp7", "Apol9a", "Tmem140", "Irf7", "Parp14", "Tgtp2", "Oas1g", "Dhx58", "Irgm2", "H2-Q6", "Herc6", "Iigp1", "Samd9l", "Atf3", "Ddx58", "Cxcl2", "Ifi204", "Gdf15", "Isg20", "Slc3a2", "Phf11d", "Oas3", "Adar", "Apol9b", "Ifi35", "Rnf213", "Ifi47", "Trim56", "Tor3a", "Mndal", "Ddx60", "H2-Q7", "Stat1", "Irgm1", "Gm4070", "Tapbp", "Pnpt1", "Enpp4", "Tlr3", "Trafd1", "H2-T10", "Trim30a", "Clic4", "Irf1", "Parp10", "Oas2", "Marchf5", "Aftph", "Zc3hav1", "Trim30d", "Eif2ak2", "H2-K1", "Ifi205", "Trex1", "Znfx1", "Ppp1r15a", "Trim25", "Nfkb2", "Il33", "Slfn2", "Parp12", "Rtp4", "Tap1", "Slfn8", "Xaf1", "Gem", "H4c9", "Ttc39c", "H2-T22", "Csrnp1", "H2aw", "Trim21", "Pnp", "Atf4", "Sqstm1", "Ifi203", "Tent5a", "Ddit3", "Lgals9", "P2rx7", "Parp9", "Arhgef2", "Relb", "Nfkb1", "Ccl2", "Daxx", "Aars", "Tnfaip3", "Irf9", "Ccnl1", "Gadd45b", "Arl14ep", "Tnip1", "Wdfy1", "Ifit2", "H2-T23", "Klf6", "Polr2g", "Chka", "Vcam1", "Tlr2", "H1f2", "Ifrd1")

# Cluster 2 from PSC.singlets.rPCA.integrated
Crabp1.CAF_Sig <- c("Hmga1b", "Spp1", "Hmga2", "Ccnd1", "Crabp1", "Mmp3", "Il1rl1", "Clu")

# Cluster 3 from PSC.singlets.rPCA.integrated
qPSC.1_Sig <- c("Dpt", "Gpnmb", "Cxcl12", "Oxct1", "Cdo1", "Cst3", "Ptn", "F3", "Mme", "Gsn", "Cd24a", "Rerg", "Lamp1", "Ftl1", "Atp1b1", "Itm2b", "Hexa", "Ctsa", "Aebp1", "Sulf2", "Maf", "Laptm4a", "mt-Nd4", "Timp2", "Lpl", "Gas6", "Ftl1-ps1", "Mfsd1", "Slc5a3", "mt-Cytb", "Lamp2", "Marcks", "mt-Nd2", "Mmp11", "Selenop", "Gabarap", "Cbx6", "Grem1", "Gstm1", "Cpe", "Phactr1", "Gns", "Adamts5", "Gaa", "Grn", "Cmtm3", "Mgp", "Dpp7", "Foxq1", "Nfia", "Cuta", "Txnip", "Col6a1", "mt-Nd1", "Uap1l1", "Zbtb20", "Cx3cl1", "S100a4", "Sox4", "Lgals3", "Sparcl1", "Cyb5a", "Tpp1", "Psap", "Atp6v0c", "Scarf2", "Pcolce", "Deptor", "Btg1", "Eid1", "Enpp2", "Setbp1", "Mrps6", "AC149090.1", "Ar", "Olfm1", "Ccn3", "Col4a1", "Enpp5", "Cygb", "Xdh", "H3f3a", "Cnpy2", "Larp6", "Gxylt2", "Nr2f1", "Amdhd2", "Ddr2", "Mfap2", "Ahnak2", "mt-Nd5", "Sfrp2", "Scpep1", "Impact", "Sipa1l2", "Ets2", "Rnase4", "Col6a2", "Thbs2", "Washc2", "Dclk1", "Col4a2", "Thbs1", "Trp53inp1", "Abi3bp", "Arid5b", "Tgfbi", "Ugcg", "Lnpep", "Ap3s1", "Amot", "Crip2", "Rin2", "Nr2f2", "Celf2", "Rbpms", "Rere", "Gfra1", "Nfix", "Klhl24", "Cdkn1b", "Qpct", "Efemp1", "Tcn2", "Vegfb", "Apoe", "Iah1", "Fam43a", "Colec12", "Myo10", "Lox", "Thra", "Dkk3", "Vgll3", "Rasl11a", "Lhfp", "Gdpd2", "Gas1", "Atg12", "Tulp4", "Cyth3", "Pmp22", "Adamts1", "Ccn1", "Serpina3n", "Igfbp4", "Serpine2", "Adgrl2", "Fhl2", "Peg10", "Zfp503", "Jade1", "Sesn3", "Fbln1", "Dlk1")
  
# Cluster 5 from PSC.singlets.rPCA.integrated
qPSC.2_Sig <- c("Alkal1", "Siglecg", "Ogn", "Adamts5", "Vegfd", "Adcyap1r1", "Adm", "Gm9780", "Nid2", "Svep1", "Myl9", "Lpl", "Sulf2", "Cavin2", "Ptx3", "Serpinb9b", "Gas6", "Fbln5", "Cryab", "Ccn3", "Prelp", "Ankrd1", "Tagln", "Amotl2", "Ptn", "Anxa3", "Pkia", "Masp1", "Cxcl12", "Pltp", "Fhl2", "F3", "Acta2", "Ephx1", "Peg10", "Plac9a", "Vgll3", "Adamts1", "Prss23", "Timp3", "Ftl1", "Man2a1", "Timp2", "Fhl1", "Thbs1", "Dpt", "Marcks", "Ahnak", "Septin11", "Fam43a", "Atp1b1", "Cnn2", "Lurap1l", "Crim1", "Basp1", "Sfrp2", "Ccng1", "Rsu1", "Ablim1", "Crip2", "Ftl1-ps1", "Vim", "Lamp1", "Igfbp6", "Nacc2", "Cdkn2a", "Bax", "Cyp1b1", "Mllt3", "Lpp", "Atf5", "Arhgap29", "Serpinb1a", "Cacna1a", "Uap1l1", "Rps27l", "Ugcg", "Gstm1", "Cyb5a", "Ryk", "S100a11", "Scrn1", "Ly6a", "Ccnd2", "Olfml2b", "Dlc1", "S100a4", "Ly6c1", "Tpm1", "Parp3", "Oxct1", "S100a6", "Tpm2", "Ei24", "Cavin1", "Mtch2", "Mn1", "Trp53inp1", "Hspa1b", "Amdhd2", "Rras2", "Lasp1", "Cuta", "Aebp1", "Rnd3", "Tle5", "Pmm1", "Fbln7")
```

```{r}
myCAF_Z_Scores <- export_Z(myCAF_Sig, c("Mono", "Co"), "PSC", c("DMSO", "DMXAA"), DNRC_long = DNRC_long_ifCAFs, Sample_md_ifCAFs)
iCAF_Z_Scores <- export_Z(iCAF_Sig, c("Mono", "Co"), "PSC", c("DMSO", "DMXAA"), DNRC_long = DNRC_long_ifCAFs, Sample_md_ifCAFs)
ifCAF_Z_Scores <- export_Z(ifCAF_Sig, c("Mono", "Co"), "PSC", c("DMSO", "DMXAA"), DNRC_long = DNRC_long_ifCAFs, Sample_md_ifCAFs)
Crabp1.CAF_Z_Scores <- export_Z(Crabp1.CAF_Sig, c("Mono", "Co"), "PSC", c("DMSO", "DMXAA"), DNRC_long = DNRC_long_ifCAFs, Sample_md_ifCAFs)
qPSC.1_Z_Scores <- export_Z(qPSC.1_Sig, c("Mono", "Co"), "PSC", c("DMSO", "DMXAA"), DNRC_long = DNRC_long_ifCAFs, Sample_md_ifCAFs)
qPSC.2.CAF_Z_Scores <- export_Z(qPSC.2_Sig, c("Mono", "Co"), "PSC", c("DMSO", "DMXAA"), DNRC_long = DNRC_long_ifCAFs, Sample_md_ifCAFs)

#write.table(myCAF_Z_Scores, "myCAF_Z_Scores_DMXAA_treatment_Mono_Co.txt", sep = "\t")
#write.table(iCAF_Z_Scores, "iCAF_Z_Scores_DMXAA_treatment_Mono_Co.txt", sep = "\t")
#write.table(ifCAF_Z_Scores, "ifCAF_Z_Scores_DMXAA_treatment_Mono_Co.txt", sep = "\t")
#write.table(Crabp1.CAF_Z_Scores, "Crabp1.CAF_Z_Scores_DMXAA_treatment_Mono_Co.txt", sep = "\t")
#write.table(qPSC.1_Z_Scores, "qPSC.1_Z_Scores_DMXAA_treatment_Mono_Co.txt", sep = "\t")
#write.table(qPSC.2.CAF_Z_Scores, "qPSC.2.CAF_Z_Scores_DMXAA_treatment_Mono_Co.txt", sep = "\t")
```


# apCAF experiment
The samples from the apCAF induction experiment (s33-s68) are selected
```{r}
Sample_md_apCAFs<-Sample_md %>% 
  filter(Sample %in% paste("s", 33:68, sep=""))
```

```{r}
RRC_long_apCAFs <- RRC_long %>%
  filter(Sample %in% Sample_md_apCAFs$Sample)
```

A wide table of raw read count table is compiled with same gene order and sample order as the metadata. The Sample name is converted to a factor to remain in the same order instead of beingg sorted by alphabetical order. 
```{r}
RRC_apCAFs<-RRC_long_apCAFs %>% 
  mutate(Sample=factor(Sample, levels = unique(RRC_long$Sample))) %>% 
  spread(key = Sample, value = RRC) 

RRC_apCAFs<-Gene_md %>% 
  select(Gene=Gene_Name) %>% 
  left_join(RRC_apCAFs, by="Gene") %>% 
  select(-EnsID)

RRC_mx_apCAFs <- RRC_apCAFs %>% 
  column_to_rownames("Gene") 
```


let's update the sample metadata to include the combined variable. We keep only the info that is useful for DE analysis.  
```{r}
Sample_md_Compvar_apCAFs <- Sample_md_apCAFs %>% 
  mutate(Condition=paste(Drug, Culture, Cell_type, MHCII_status, sep="_")) %>% 
  select(Sample, Run, Biological.Replicate, Drug, Culture, Cell_type, Condition)
```


```{r}
dds_apCAFs <- DESeqDataSetFromMatrix(countData = RRC_mx_apCAFs,
                              colData = Sample_md_Compvar_apCAFs,
                              design= ~ Run + Condition)

dds_apCAFs <- DESeq(dds_apCAFs)
```

The DEseq-normalized read count (DNRC) is extracted 
```{r}
DNRC_apCAFs <- counts(dds_apCAFs, normalized=TRUE) %>% 
  as.data.frame() %>% 
  rownames_to_column("Gene")

DNRC_long_apCAFs <- DNRC_apCAFs %>% 
  gather(key = Sample, value = DNRC, -Gene)
```

```{r}
FC_qvals_apCAFs <- Qs(dds_apCAFs, "Condition", "IFNγ_96hr_Mono_PSC_MHCIIPositive", "Ctrl_Mono_PSC_MHCII_Negative") %>% 
  bind_rows(Qs(dds_apCAFs, "Condition", "IFNγ_Co_PSC_MHCIIPositive", "Ctrl_Co_PSC_MHCII_Negative")) %>% 
  bind_rows(Qs(dds_apCAFs, "Condition", "IFNγ_Co_PSC_MHCII_Negative", "Ctrl_Co_PSC_MHCII_Negative")) %>% 
  bind_rows(Qs(dds_apCAFs, "Condition", "IFNγ_96hr_Co_PSC_MHCIIPositive", "Ctrl_Co_PSC_MHCII_Negative")) %>% 
  bind_rows(Qs(dds_apCAFs, "Condition", "IFNγ_96hr_Co_PSC_MHCII_Negative", "Ctrl_Co_PSC_MHCII_Negative")) %>% 
  bind_rows(Qs(dds_apCAFs, "Condition", "T-Cells_Mono_PSC_MHCIIPositive", "Ctrl_Mono_PSC_MHCII_Negative"))
```

The mean DNRC per group is introduced
```{r}
group_DNRC_apCAFs <-DNRC_long_apCAFs %>% 
  left_join(Sample_md_Compvar_apCAFs, by = "Sample") %>% 
  group_by(Gene, Condition) %>% 
  summarise(mean_DNRC=round(mean(DNRC), 3)) 

Means_FC_qvals_apCAFs <- FC_qvals_apCAFs %>% 
  left_join(group_DNRC_apCAFs, by=c("Gene","ctrl_group" = "Condition")) %>% 
  dplyr::rename(ctrl_mean_DNRC=mean_DNRC) %>% 
  left_join(group_DNRC_apCAFs, by=c("Gene","treat_group" = "Condition")) %>% 
  dplyr::rename(treat_mean_DNRC=mean_DNRC) %>% 
  mutate(FC=round(FC, 2)) %>% 
  drop_na()
```


The DE genes are extracted with lenient cutoff for plotting purposes
```{r}
LenSel_Means_FC_qvals_apCAFs <-Means_FC_qvals_apCAFs %>% 
  filter(map2(ctrl_mean_DNRC, treat_mean_DNRC, max)>3) %>% 
  mutate(FC = FC+0.01) %>%
  mutate(logFC=log(FC)) %>% 
  filter(abs(logFC)>0.1) %>% 
  filter(padj<0.8) %>% 
  arrange(desc(FC))
```


# Drug-modulated genes per condition
DE genes are selected that are modulated by DMXAA or IFNg
```{r}
DMXAA_Mono_PSC_gr_comp_qvals <- LenSel_Means_FC_qvals_ifCAFs %>% 
    filter(treat_group== "DMXAA_Mono_PSC") %>% 
    filter(ctrl_group== "DMSO_Mono_PSC") %>% 
    filter(ctrl_mean_DNRC >40 | treat_mean_DNRC >40) %>%
    mutate(sig=ifelse(padj<0.05 & abs(log(FC))>0.405, "S", "NS")) %>% 
    filter(sig=="S") %>%
    column_to_rownames("Gene")

IFNy_MHCII_pos_Mono_PSC_gr_comp_qvals <- LenSel_Means_FC_qvals_apCAFs %>% 
    filter(treat_group=="IFNγ_96hr_Mono_PSC_MHCIIPositive") %>% 
    filter(ctrl_group== "Ctrl_Mono_PSC_MHCII_Negative") %>% 
    filter(ctrl_mean_DNRC >50 | treat_mean_DNRC >50) %>%
    mutate(sig=ifelse(padj<0.05 & abs(log(FC))>0.405, "S", "NS")) %>% 
    filter(sig=="S") %>%
    column_to_rownames("Gene")

#write.table(DMXAA_Mono_PSC_gr_comp_qvals, "DMXAA_Mono_PSC_DEGs.txt", sep = "\t")
#write.table(IFNy_MHCII_pos_Mono_PSC_gr_comp_qvals, "IFNy_MHCII_pos_Mono_PSC_DEGs.txt", sep = "\t")
```


# Gene set enrichment analysis
clusterProfiler is used for gene set enrichment analysis
```{r}
library(clusterProfiler)  # version 4.2.2
library(org.Mm.eg.db)
library(ReactomePA)
library(msigdbr)  # version 7.5.1
library(stringr)
library(forcats)
library(enrichplot)  # version 1.14.2
library(ggstance)
```

A function for the GSEA analysis  GSEA based fold-change. It needs an input dataframe with DE genes including a column named `logFC` with the gene names as rownames. The function can be run in mode "reactome", "KEGG" , "GO", or "custom" depending on the desired analysis. In custom mode, the gene set must be provided. 
The enrichment results are output as a dataframe
```{r}
Get_enrichment_DF <-function(in_df, mode="reactome", custom_gene_set=NA){
  FC_df<-in_df %>% 
    rownames_to_column("Gene") %>% 
    dplyr::select(Gene, logFC)
  
  gen_ID_key<-AnnotationDbi::select(org.Mm.eg.db, columns = c("ENTREZID", "SYMBOL"), keys = FC_df$Gene, keytype = "SYMBOL")
  
  fc_vector<-FC_df %>% 
    left_join(gen_ID_key, by=c("Gene"="SYMBOL")) %>% 
    drop_na() %>% 
    dplyr::select(ENTREZID, logFC) %>% 
    arrange(desc(logFC)) %>% 
    deframe()
  
  set.seed(1234)
  
  if (mode=="reactome"){
    gse <- gsePathway(geneList = fc_vector, organism = 'mouse', minGSSize = 10, pvalueCutoff = 0.05, pAdjustMethod = "BH", verbose = FALSE, seed = T)}
  
  if (mode=="KEGG"){
    gse <- gseKEGG(geneList = fc_vector, organism = 'mmu', minGSSize = 10, pvalueCutoff = 0.05, pAdjustMethod = "BH", verbose = FALSE, seed = T)}
  
  if (mode=="GO"){
    gse <- gseGO(geneList = fc_vector, OrgDb = org.Mm.eg.db, ont = "CC", minGSSize = 1, maxGSSize = 5000, pvalueCutoff = 0.05, pAdjustMethod = "BH", verbose = FALSE, seed = T)}
  
  if (mode=="custom"){
    gse <- GSEA(geneList = fc_vector, TERM2GENE = custom_gene_set, minGSSize = 1, maxGSSize = 5000, pvalueCutoff = 0.05, pAdjustMethod = "BH", verbose = FALSE, seed = T)}
  
  gse_df<-gse@result %>%
    filter(p.adjust < 0.05)
  
  return(gse_df)
}
```

Specific gene set annotations are selected for custom ORA analysis
H: hallmark gene sets
C1: positional gene sets
C2: curated gene sets
C3: motif gene sets
C4: computational gene sets
C5: GO gene sets
C6: oncogenic signatures
C7: immunologic signatures
```{r}
H_Geneset <- msigdbr(species = "Mus musculus", category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)
C1_Geneset <- msigdbr(species = "Mus musculus", category = "C1") %>% 
  dplyr::select(gs_name, entrez_gene)
C2_Geneset <- msigdbr(species = "Mus musculus", category = "C2") %>% 
  dplyr::select(gs_name, entrez_gene)
C3_Geneset <- msigdbr(species = "Mus musculus", category = "C3") %>% 
  dplyr::select(gs_name, entrez_gene)
C4_Geneset <- msigdbr(species = "Mus musculus", category = "C4") %>% 
  dplyr::select(gs_name, entrez_gene)
C5_Geneset <- msigdbr(species = "Mus musculus", category = "C5") %>% 
  dplyr::select(gs_name, entrez_gene)
C6_Geneset <- msigdbr(species = "Mus musculus", category = "C6") %>% 
  dplyr::select(gs_name, entrez_gene)
C7_Geneset <- msigdbr(species = "Mus musculus", category = "C7") %>% 
  dplyr::select(gs_name, entrez_gene)
```


Pathways enriched in DMXAA treated samples
```{r}
Mono.PSCs.DMXAA.Reactome.GSEA <- Get_enrichment_DF(in_df = DMXAA_Mono_PSC_gr_comp_qvals, mode = "reactome")
Mono.PSCs.DMXAA.KEGG.GSEA <- Get_enrichment_DF(in_df = DMXAA_Mono_PSC_gr_comp_qvals, mode = "KEGG")
Mono.PSCs.DMXAA.H.GSEA <- Get_enrichment_DF(in_df = DMXAA_Mono_PSC_gr_comp_qvals, mode = "custom", H_Geneset)
Mono.PSCs.DMXAA.GO.GSEA <- Get_enrichment_DF(in_df = DMXAA_Mono_PSC_gr_comp_qvals, mode = "GO")

Mono.PSCs.DMXAA.GSEA.Summary <- rbind(Mono.PSCs.DMXAA.Reactome.GSEA, Mono.PSCs.DMXAA.KEGG.GSEA, Mono.PSCs.DMXAA.H.GSEA, Mono.PSCs.DMXAA.GO.GSEA)
```

Pathways enriched in IFNg treated samples
```{r}
Mono.PSCs.IFNy.Reactome.GSEA <- Get_enrichment_DF(in_df = IFNy_MHCII_pos_Mono_PSC_gr_comp_qvals, mode = "reactome")
Mono.PSCs.IFNy.KEGG.GSEA <- Get_enrichment_DF(in_df = IFNy_MHCII_pos_Mono_PSC_gr_comp_qvals, mode = "KEGG")
Mono.PSCs.IFNy.H.GSEA <- Get_enrichment_DF(in_df = IFNy_MHCII_pos_Mono_PSC_gr_comp_qvals, mode = "custom", H_Geneset)
Mono.PSCs.IFNy.GO.GSEA <- Get_enrichment_DF(in_df = IFNy_MHCII_pos_Mono_PSC_gr_comp_qvals, mode = "GO")

Mono.PSCs.IFNy.GSEA.Summary <- rbind(Mono.PSCs.IFNy.Reactome.GSEA, Mono.PSCs.IFNy.KEGG.GSEA, Mono.PSCs.IFNy.H.GSEA, Mono.PSCs.IFNy.GO.GSEA)
```

Common and unique pathways are distinguished
```{r}
Common.Pathways.Mono.PSC.IFNy.DMXAA <- Mono.PSCs.IFNy.GSEA.Summary %>%
dplyr::filter(Description %in% Mono.PSCs.DMXAA.GSEA.Summary$Description)

Unique.Pathways.Mono.PSC.IFNy.vs.DMXAA <- Mono.PSCs.IFNy.GSEA.Summary %>%
dplyr::filter(!Description %in% Mono.PSCs.DMXAA.GSEA.Summary$Description)

Unique.Pathways.Mono.PSC.DMXAA.vs.IFNy <- Mono.PSCs.DMXAA.GSEA.Summary %>%
dplyr::filter(!Description %in% Mono.PSCs.IFNy.GSEA.Summary$Description)
```

```{r}
DMXAA.IFNy.selected.pathways <- c("HALLMARK_INTERFERON_GAMMA_RESPONSE", "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_INFLAMMATORY_RESPONSE", "Toll-like receptor signaling pathway - Mus musculus (house mouse)", "Cytosolic DNA-sensing pathway - Mus musculus (house mouse)", "Antigen processing and presentation", "Antigen processing-Cross presentation", "MHC class II protein complex", "MHC protein complex", "Antigen Presentation: Folding, assembly and peptide loading of class I MHC")
```

```{r}
selected.DMXAA.Mono.PSC.GSEA.Summary <- Mono.PSCs.DMXAA.GSEA.Summary %>%
  dplyr::filter(Description %in% DMXAA.IFNy.selected.pathways) %>%
  dplyr::mutate(Treatment = "DMXAA")

selected.IFNy.Mono.PSC.GSEA.Summary <- Mono.PSCs.IFNy.GSEA.Summary %>%
  dplyr::filter(Description %in% DMXAA.IFNy.selected.pathways) %>%
  dplyr::mutate(Treatment = "IFNy")

selected.DMXAA.IFNy.Mono.PSC.GSEA.Summary <- rbind(selected.DMXAA.Mono.PSC.GSEA.Summary, selected.IFNy.Mono.PSC.GSEA.Summary)

selected.DMXAA.IFNy.Mono.PSC.GSEA.Summary$Description <- factor(selected.DMXAA.IFNy.Mono.PSC.GSEA.Summary$Description, levels = c("HALLMARK_INTERFERON_GAMMA_RESPONSE", "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_INFLAMMATORY_RESPONSE", "Toll-like receptor signaling pathway - Mus musculus (house mouse)", "Cytosolic DNA-sensing pathway - Mus musculus (house mouse)", "Antigen processing and presentation", "Antigen processing-Cross presentation", "MHC class II protein complex", "MHC protein complex", "Antigen Presentation: Folding, assembly and peptide loading of class I MHC"))
```


# Figure S8N
```{r}
Fig_S8N_plot <- ggplot(selected.DMXAA.IFNy.Mono.PSC.GSEA.Summary, aes(x = Treatment, y = Description)) + 
               geom_point(aes(color = NES, size = setSize)) +
               theme_bw() +
        scale_colour_gradient(limits=c(0, 2.5), low="white", high = "red") +
        #scale_size(range = c(2,8)) +
        ylab(NULL) +
        ggtitle("") +
  theme(axis.text.x = element_text(angle = 45, hjust=1, size = 10), title = element_text(size = 10), axis.text.y = element_text(size = 10)) +
  scale_y_discrete( labels = function(x) str_wrap(x, width = 50))

if (!dir.exists("Figure_S8")) {dir.create("Figure_S8")}
ggsave(plot = Fig_S8N_plot, "Figure_S8N_DMXAAA_IFNy_Mono_PSCs_selected_GSEA_pathways.pdf", width = 6, height = 5, units = "in", path = "Figure_S8/")
```

# Comparison of TF activity scores 
Transcription factor activities were predicted using ISMARA. The activity and delta value tables were derived for the PSC monocultures treated with DMXAA or IFNg (compared to vehivcle control). The ISMARA activity and delta tables are available on ArrayExpress with accession number E-MTAB-14943. 
Transcription factor activity scores are compared between DMXAA and IFNg inductions to find common, specific and divergent pathways.

The activity and delta tables are loaded
```{r}
activity_table_Mono_PSC_IFNy <- read.table("ISMARA_activity_table_Mono_PSC_IFNy.txt")
delta_table_Mono_PSC_IFNy <- read.table("ISMARA_delta_table_Mono_PSC_IFNy.txt")
activity_table_Mono_PSC_DMXAA <- read.table("ISMARA_activity_table_Mono_PSC_DMXAA.txt")
delta_table_Mono_PSC_DMXAA <- read.table("ISMARA_delta_table_Mono_PSC_DMXAA.txt")
```


```{r}
plot_DATFs_between_conditions <- function(in_activity_table_1, in_delta_table_1, in_condition_1, in_activity_table_2, in_delta_table_2, in_condition_2, in_group_1, in_group_2, xlabel, ylabel) {
Activity.table.Co.T.1 <- in_activity_table_1
Delta.table.Co.T.1 <- in_delta_table_1

Delta.table.Co.T.1 <- Delta.table.Co.T.1 %>%
  t()  %>%
  as.data.frame() %>%
  dplyr::select(c(Vehicle, in_condition_1)) %>%
  rename(Delta.Score.1 = in_condition_1) %>%
  rename(Vehicle.Delta.Score = Vehicle) %>%
  rownames_to_column("motif")

Activity.table.Co.T.1 <- Activity.table.Co.T.1 %>%
  t()  %>%
  as.data.frame() %>%
  dplyr::select(c(Vehicle, in_condition_1)) %>%
  rename(Activity.Score.1 = in_condition_1) %>%
  rename(Vehicle.Activity.Score = Vehicle) %>%
  rownames_to_column("motif")

Z.Scoring.1 <- Activity.table.Co.T.1 %>%
  left_join(Delta.table.Co.T.1, by = "motif") %>%
  mutate(Vehicle.Activity.over.delta = Vehicle.Activity.Score/Vehicle.Delta.Score) %>%
  mutate(Vehicle.Activity.over.delta = Vehicle.Activity.over.delta^2) %>%
  mutate(Activity.over.delta.1 = Activity.Score.1/Delta.Score.1) %>%
  mutate(Activity.over.delta.1 = Activity.over.delta.1^2) %>%
  mutate(Sum.Activity.over.delta.1 = Activity.over.delta.1 + Vehicle.Activity.over.delta) %>%
  mutate(Sum.Activity.over.delta.1 = Sum.Activity.over.delta.1/2) %>%
  mutate(Z.Score.1 = sqrt(Sum.Activity.over.delta.1)) %>%
  dplyr::select(c(motif, Z.Score.1, Vehicle.Activity.Score, Activity.Score.1))

Activity.table.Co.T.2 <- in_activity_table_2
Delta.table.Co.T.2 <- in_delta_table_2

Delta.table.Co.T.2 <- Delta.table.Co.T.2 %>%
  t()  %>%
  as.data.frame() %>%
  dplyr::select(c(Vehicle, in_condition_2)) %>%
  rename(Delta.Score.2 = in_condition_2) %>%
  rename(Vehicle.Delta.Score = Vehicle) %>%
  rownames_to_column("motif")

Activity.table.Co.T.2 <- Activity.table.Co.T.2 %>%
  t()  %>%
  as.data.frame() %>%
  dplyr::select(c(Vehicle, in_condition_2)) %>%
  rename(Activity.Score.2 = in_condition_2) %>%
  rename(Vehicle.Activity.Score = Vehicle) %>%
  rownames_to_column("motif")

Z.Scoring.2 <- Activity.table.Co.T.2 %>%
  left_join(Delta.table.Co.T.2, by = "motif") %>%
  mutate(Vehicle.Activity.over.delta = Vehicle.Activity.Score/Vehicle.Delta.Score) %>%
  mutate(Vehicle.Activity.over.delta = Vehicle.Activity.over.delta^2) %>%
  mutate(Activity.over.delta.2 = Activity.Score.2/Delta.Score.2) %>%
  mutate(Activity.over.delta.2 = Activity.over.delta.2^2) %>%
  mutate(Sum.Activity.over.delta.2 = Activity.over.delta.2 + Vehicle.Activity.over.delta) %>%
  mutate(Sum.Activity.over.delta.2 = Sum.Activity.over.delta.2/2) %>%
  mutate(Z.Score.2 = sqrt(Sum.Activity.over.delta.2)) %>%
  dplyr::select(c(motif, Z.Score.2, Activity.Score.2))

Z.Scoring <- Z.Scoring.1 %>% 
  left_join(Z.Scoring.2, by = "motif") 
Z.Scoring <- Z.Scoring %>%
  mutate(tag=ifelse(Z.Score.1 > 2 & Z.Score.2 < 2, in_group_1, NA)) %>%
  mutate(tag=ifelse(Z.Score.1 < 2 & Z.Score.2 < 2, "NS", tag)) %>%
  mutate(tag=ifelse(Z.Score.2 > 2 & Z.Score.1 < 2, in_group_2, tag)) %>%
  mutate(tag=ifelse(Z.Score.2 > 2 & Z.Score.1 > 2, "Common DATFs", tag)) %>%
  mutate(Activity.Diff.1 = Activity.Score.1 - Vehicle.Activity.Score) %>%
  mutate(Activity.Diff.2 = Activity.Score.2 - Vehicle.Activity.Score) %>%
  mutate(tag=ifelse(Z.Score.1 >2 & Z.Score.2 > 2 & Activity.Diff.1 > 0 & Activity.Diff.2 < 0 | Z.Score.1 >2 & Z.Score.2 > 2 & Activity.Diff.1 < 0 & Activity.Diff.2 > 0, "Divergent_DATFs", tag)) %>%
  drop_na()  %>%
  mutate(Z.Diff = Z.Score.1 - Z.Score.2) %>%
  mutate(Z.Product = Z.Score.1 * Z.Score.2)

Condition.1.Top.specific.DATFs <- Z.Scoring %>%
  filter(tag %in% in_group_1) %>%
  top_n(3, Z.Diff)

Condition.2.Top.specific.DATFs <- Z.Scoring %>%
  filter(tag %in% in_group_2) %>%
  top_n(-3, Z.Diff)

Top_Common_DATFS <- Z.Scoring %>%
  filter(tag %in% "Common DATFs") %>%
  top_n(4, Z.Product)

Selected.DATFS <- rbind(Condition.1.Top.specific.DATFs, Condition.2.Top.specific.DATFs, Top_Common_DATFS)

cols <- c(in_group_1 = "#ffad73", in_group_2 = "#26b3ff", "Common DATFs" = "#76b96c",  "NS" = "grey") 
sizes <- c("in_group_1" = 2, "in_group_2" = 2, "Common DATFs" =2, "ns" = 1) 
alphas <- c("in_group_1" = 1, "in_group_2" = 1, "Common DATFs" =1, "ns" = 0.5)

ggplot(Z.Scoring, aes(x=Z.Score.1, y=Z.Score.2)) + 
  geom_point(aes(colour = tag),
             shape = 16,
             size = 1) +
  geom_point(data = Condition.1.Top.specific.DATFs,
             shape = 21,
             size = 3, 
             fill = "#ffad73", 
             colour = "black") +
  geom_point(data = Condition.2.Top.specific.DATFs,
             shape = 21,
             size = 2, 
             fill = "#26b3ff", 
             colour = "black") +
  geom_point(data = Top_Common_DATFS,
             shape = 21,
             size = 2, 
             fill = "#76b96c", 
             colour = "black") + 
  geom_hline(yintercept = 2,
             linetype = "dashed") + 
  geom_vline(xintercept = 2,
             linetype = "dashed") +
  geom_text_repel(data = Selected.DATFS, aes(label = motif), force = 2, nudge_y = 1, show.legend = FALSE, box.padding = 0.5) +
  xlab(xlabel) +
  ylab(ylabel) +
  theme_bw() + # Select theme with a white background  
  theme(panel.border = element_rect(colour = "black", fill = NA, size= 0.5),    
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank())
}

```

```{r}
plot.IFNy.vs.DMXAA <- plot_DATFs_between_conditions(in_activity_table_1 = activity_table_Mono_PSC_IFNy, in_delta_table_1 = delta_table_Mono_PSC_IFNy, in_condition_1 = "IFNy", in_activity_table_2 = activity_table_Mono_PSC_DMXAA, in_delta_table_2 = delta_table_Mono_PSC_DMXAA, in_condition_2 = "DMXAA", in_group_1 = "IFNy Specific DATFs", in_group_2 = "DMXAA Specific DATFs", xlabel = "Z-Value IFNy", ylabel = "Z-Value DMXAA")
```

# Figure S8O
```{r}
Fig_S8O_plot <- plot.IFNy.vs.DMXAA +
  scale_color_manual( values =c("IFNy Specific DATFs" = "#ffad73", "DMXAA Specific DATFs" = "#26b3ff", "Common DATFs" = "#76b96c",  "NS" = "grey"))

if (!dir.exists("Figure_S8")) {dir.create("Figure_S8")}
ggsave(plot = Fig_S8O_plot, "Figure_S8O_DMXAA_IFNy_Mono_PSCs_DATFs.pdf", width = 6, height = 6, units = "in", path = "Figure_S8/")
```

