---
title: "PSC_Time_course_Figure_4_S7"
author: "Joshua"
date: '2023-01-05'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary
The following code describes the analysis of mouse single cell RNAseq data from a time-course of co-cultures of pancreatic stellate cells (PSC) in the presence of tumour organids. This code was used to produce figures 4 and S7 of Cumming et al, 2025. 

The input data is the raw UMI count from the single cell RNA-seq data generated with ddSeq for 15 samples of PSC extracted from co-cultures across a time-course from day 0 to day 6. The raw UMI count is available on ArrayExpress with accession number E-MTAB-14945. 

The code to generate the raw umi count is described in an Rmd called `PSC_timecourse_15s_scSeq_fastq_to_count.Rmd`

The Sample metadata required was prepared manually and is available in the file `PSC_timecourse_15s_scSeq_Sample_metadata.txt`

# Load libraries
```{r}
library(tidyverse)
library(Seurat)  # version 4.1.0

library(DoubletFinder)  # version 2.0.3
library(slingshot)  # version 2.8.0
library(DelayedMatrixStats)

library(ggplotify)
library(ggpubr)
library(viridis)
library(cowplot)
library(decoupleR)  # version 2.0.1
library(pheatmap)  # version 1.0.12
library(RColorBrewer)

library(clusterProfiler)  # version 4.2.2
library(org.Mm.eg.db)
library(ReactomePA)
library(msigdbr)  # version 7.5.1

library(SCPA)  # version 1.5.3

library(scales)
library(scran)
library(igraph)
```


# Loading of data and metadata
The raw UMI count is loaded to a seurat object
```{r}
Timecourse_15s_SO <- CreateSeuratObject(counts = Read10X_h5(filename = "PSC_timecourse_15s_Raw_UMI_Count.h5"))
```

The sample metadata is loaded and integrated in the Seurat object
```{r}
Sample_md<-tibble(Sample=paste("s", 1:15, sep=""), Time.point=c("Day.0", "Day.1", "Day.3", "Day.4", "Day.5", "Day.5", "Day.2", "Day.4", "Day.2", "Day.3", "Day.6", "Day.1", "Day.1", "Day.0", "Day.6"), Replicate=c("1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1"), Batch=c("B1", "B1", "B1", "B2", "B2", "B2", "B3", "B3", "B4", "B4", "B4", "B5", "B5", "B6", "B6"))

Timecourse_15s_SO@meta.data<-Timecourse_15s_SO@meta.data %>% 
  rownames_to_column("Cell") %>% 
  mutate(Sample=orig.ident) %>% 
  left_join(Sample_md, by="Sample") %>% 
  mutate(Sample=factor(Sample, levels=paste("s", as.character(1:15), sep=""))) %>% 
  column_to_rownames("Cell")
```

# Quality control metrics
The percentage of mitochondrial reads is calculated. The UMI count, number of genes and percentage mitochondrial reads are observed.
```{r}
Timecourse_15s_SO[["percent.mt"]] <- PercentageFeatureSet(Timecourse_15s_SO, pattern = "^mt-")

VlnPlot(Timecourse_15s_SO, features = c("nFeature_RNA"), split.by = "Day", group.by = "Sample")
VlnPlot(Timecourse_15s_SO, features = c("nCount_RNA"), split.by = "Day", group.by = "Sample")
VlnPlot(Timecourse_15s_SO, features = c("percent.mt"), split.by = "Day", group.by = "Sample")
```



# Normalization and dimensionality reduction
The seurat object is split by sample into 15 individual objects 
```{r}
Timecourse_15SO_list<-SplitObject(Timecourse_15s_SO, split.by = "Sample")
```

The normalization and dimensionality reduction is run in the same way on all 15 objects. Clusters and UMAP are also calculated with same parameters
```{r}
for (i in 1:15) {
  Timecourse_15SO_list[[i]] <- NormalizeData(Timecourse_15SO_list[[i]], normalization.method = "LogNormalize", scale.factor = 10000)
  Timecourse_15SO_list[[i]] <- FindVariableFeatures(Timecourse_15SO_list[[i]], selection.method = "vst", nfeatures = 2000)
  Timecourse_15SO_list[[i]] <- ScaleData(Timecourse_15SO_list[[i]], features = rownames(Timecourse_15SO_list[[i]]))
  Timecourse_15SO_list[[i]] <- RunPCA(Timecourse_15SO_list[[i]], features = VariableFeatures(object = Timecourse_15SO_list[[i]]))
  Timecourse_15SO_list[[i]] <- FindNeighbors(Timecourse_15SO_list[[i]], dims = 1:20)
  Timecourse_15SO_list[[i]] <- FindClusters(Timecourse_15SO_list[[i]], resolution = 1.2)
  Timecourse_15SO_list[[i]] <- RunUMAP(Timecourse_15SO_list[[i]], dims = 1:20)
}

S1<-Timecourse_15SO_list[[1]]
S2<-Timecourse_15SO_list[[2]]
S3<-Timecourse_15SO_list[[3]]
S4<-Timecourse_15SO_list[[4]]
S5<-Timecourse_15SO_list[[5]]
S6<-Timecourse_15SO_list[[6]]
S7<-Timecourse_15SO_list[[7]]
S8<-Timecourse_15SO_list[[8]]
S9<-Timecourse_15SO_list[[9]]
S10<-Timecourse_15SO_list[[10]]
S11<-Timecourse_15SO_list[[11]]
S12<-Timecourse_15SO_list[[12]]
S13<-Timecourse_15SO_list[[13]]
S14<-Timecourse_15SO_list[[14]]
S15<-Timecourse_15SO_list[[15]]

```

The quality metric plots are observed
```{r}
VlnPlot(S1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S4, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S5, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S6, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S7, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S8, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S9, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S10, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S11, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S12, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S13, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S14, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S15, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

Some clusters represent poor-quality cells and are removed. The suerat object is subsetted to remove the vile cluster(s) and re-normalized. 
```{r}
S2 <- subset(S2, idents = c("0", "1", "2"))  # remove cl. 3
S3 <- subset(S3, idents = c("0", "1", "2"))  # remove cl. 3
S9 <- subset(S9, idents = c("0", "1", "3"))  # remove cl. 2
S10 <- subset(S10, idents = c("0", "1", "2"))  # remove cl. 3
S11 <- subset(S11, idents = c("0", "2", "3", "4"))  # remove cl. 1
S14 <- subset(S14, idents = c("0", "1", "2", "3"))  # remove cl. 4

Timecourse_6SO_list<-c(S2,S3,S9,S10,S11,S14)

for (i in 1:6) {
  Timecourse_6SO_list[[i]] <- NormalizeData(Timecourse_6SO_list[[i]], normalization.method = "LogNormalize", scale.factor = 10000)
  Timecourse_6SO_list[[i]] <- FindVariableFeatures(Timecourse_6SO_list[[i]], selection.method = "vst", nfeatures = 2000)
  Timecourse_6SO_list[[i]] <- ScaleData(Timecourse_6SO_list[[i]], features = rownames(Timecourse_6SO_list[[i]]))
  Timecourse_6SO_list[[i]] <- RunPCA(Timecourse_6SO_list[[i]], features = VariableFeatures(object = Timecourse_6SO_list[[i]]))
  Timecourse_6SO_list[[i]] <- FindNeighbors(Timecourse_6SO_list[[i]], dims = 1:20)
  Timecourse_6SO_list[[i]] <- FindClusters(Timecourse_6SO_list[[i]], resolution = 0.8)
  Timecourse_6SO_list[[i]] <- RunUMAP(Timecourse_6SO_list[[i]], dims = 1:20)
}

S2<-Timecourse_6SO_list[[1]]
S3<-Timecourse_6SO_list[[2]]
S9<-Timecourse_6SO_list[[3]]
S10<-Timecourse_6SO_list[[4]]
S11<-Timecourse_6SO_list[[5]]
S14<-Timecourse_6SO_list[[6]]

VlnPlot(S2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S9, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S10, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S11, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(S14, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```



# Doublet removal
The cell doublets are removed with DoubletFinder separately on each sample. Note: this was preformed with DoubletFinder version 2.0.3. Newer versions will give slightly different results.
```{r}
sweep.res.S1 <- paramSweep_v3(S1) 
sweep.res.S1 <- summarizeSweep(sweep.res.S1, GT = FALSE) 
bcmvn.S1 <- find.pK(sweep.res.S1)
nExp <- round(ncol(S1) * 0.05)  # expected doublets
S1 <- doubletFinder_v3(S1, pN = 0.25, pK = 0.09, nExp = nExp, PCs = 1:10)
DF.name = colnames(S1@meta.data)[grepl("DF.classification", colnames(S1@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S1, group.by = "ident") + NoAxes(),
    DimPlot(S1, group.by = DF.name) + NoAxes())
VlnPlot(S1, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S1_singlets = S1[, S1@meta.data[, DF.name] == "Singlet"]
DimPlot(S1_singlets, reduction = "umap")
VlnPlot(S1_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S1_singlets, "S1_singlets.rds")
```

```{r}
sweep.res.S2 <- paramSweep_v3(S2) 
sweep.res.S2 <- summarizeSweep(sweep.res.S2, GT = FALSE) 
bcmvn.S2 <- find.pK(sweep.res.S2)
nExp <- round(ncol(S2) * 0.05)  # expected doublets
S2 <- doubletFinder_v3(S2, pN = 0.25, pK = 0.25, nExp = nExp, PCs = 1:10)
DF.name = colnames(S2@meta.data)[grepl("DF.classification", colnames(S2@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S2, group.by = "ident") + NoAxes(),
    DimPlot(S2, group.by = DF.name) + NoAxes())
VlnPlot(S2, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S2_singlets = S2[, S2@meta.data[, DF.name] == "Singlet"]
DimPlot(S2_singlets, reduction = "umap")
VlnPlot(S2_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S2_singlets, "S2_singlets.rds")
```

```{r}
sweep.res.S3 <- paramSweep_v3(S3) 
sweep.res.S3 <- summarizeSweep(sweep.res.S3, GT = FALSE) 
bcmvn.S3 <- find.pK(sweep.res.S3)
nExp <- round(ncol(S3) * 0.05)  # expected doublets
S3 <- doubletFinder_v3(S3, pN = 0.25, pK = 0.2, nExp = nExp, PCs = 1:10)
DF.name = colnames(S3@meta.data)[grepl("DF.classification", colnames(S3@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S3, group.by = "ident") + NoAxes(),
    DimPlot(S3, group.by = DF.name) + NoAxes())
VlnPlot(S3, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S3_singlets = S3[, S3@meta.data[, DF.name] == "Singlet"]
DimPlot(S3_singlets, reduction = "umap")
VlnPlot(S3_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S3_singlets, "S3_singlets.rds")
```

```{r}
sweep.res.S4 <- paramSweep_v3(S4) 
sweep.res.S4 <- summarizeSweep(sweep.res.S4, GT = FALSE) 
bcmvn.S4 <- find.pK(sweep.res.S4)
nExp <- round(ncol(S4) * 0.05)  # expected doublets
S4 <- doubletFinder_v3(S4, pN = 0.25, pK = 0.12, nExp = nExp, PCs = 1:10)
DF.name = colnames(S4@meta.data)[grepl("DF.classification", colnames(S4@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S4, group.by = "ident") + NoAxes(),
    DimPlot(S4, group.by = DF.name) + NoAxes())
VlnPlot(S4, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S4_singlets = S4[, S4@meta.data[, DF.name] == "Singlet"]
DimPlot(S4_singlets, reduction = "umap")
VlnPlot(S4_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S4_singlets, "S4_singlets.rds")
```

```{r}
sweep.res.S5 <- paramSweep_v3(S5) 
sweep.res.S5 <- summarizeSweep(sweep.res.S5, GT = FALSE) 
bcmvn.S5 <- find.pK(sweep.res.S5)
nExp <- round(ncol(S5) * 0.05)  # expected doublets
S5 <- doubletFinder_v3(S5, pN = 0.25, pK = 0.21, nExp = nExp, PCs = 1:10)
DF.name = colnames(S5@meta.data)[grepl("DF.classification", colnames(S5@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S5, group.by = "ident") + NoAxes(),
    DimPlot(S5, group.by = DF.name) + NoAxes())
VlnPlot(S5, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S5_singlets = S5[, S5@meta.data[, DF.name] == "Singlet"]
DimPlot(S5_singlets, reduction = "umap")
VlnPlot(S5_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S5_singlets, "S5_singlets.rds")
```

```{r}
sweep.res.S6 <- paramSweep_v3(S6) 
sweep.res.S6 <- summarizeSweep(sweep.res.S6, GT = FALSE) 
bcmvn.S6 <- find.pK(sweep.res.S6)
nExp <- round(ncol(S6) * 0.05)  # expected doublets
S6 <- doubletFinder_v3(S6, pN = 0.25, pK = 0.12, nExp = nExp, PCs = 1:10)
DF.name = colnames(S6@meta.data)[grepl("DF.classification", colnames(S6@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S6, group.by = "ident") + NoAxes(),
    DimPlot(S6, group.by = DF.name) + NoAxes())
VlnPlot(S6, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S6_singlets = S6[, S6@meta.data[, DF.name] == "Singlet"]
DimPlot(S6_singlets, reduction = "umap")
VlnPlot(S6_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S6_singlets, "S6_singlets.rds")
```

```{r}
sweep.res.S7 <- paramSweep_v3(S7) 
sweep.res.S7 <- summarizeSweep(sweep.res.S7, GT = FALSE) 
bcmvn.S7 <- find.pK(sweep.res.S7)
nExp <- round(ncol(S7) * 0.05)  # expected doublets
S7 <- doubletFinder_v3(S7, pN = 0.25, pK = 0.18, nExp = nExp, PCs = 1:10)
DF.name = colnames(S7@meta.data)[grepl("DF.classification", colnames(S7@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S7, group.by = "ident") + NoAxes(),
    DimPlot(S7, group.by = DF.name) + NoAxes())
VlnPlot(S7, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S7_singlets = S7[, S7@meta.data[, DF.name] == "Singlet"]
DimPlot(S7_singlets, reduction = "umap")
VlnPlot(S7_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S7_singlets, "S7_singlets.rds")
```

```{r}
sweep.res.S8 <- paramSweep_v3(S8) 
sweep.res.S8 <- summarizeSweep(sweep.res.S8, GT = FALSE) 
bcmvn.S8 <- find.pK(sweep.res.S8)
nExp <- round(ncol(S8) * 0.05)  # expected doublets
S8 <- doubletFinder_v3(S8, pN = 0.25, pK = 0.15, nExp = nExp, PCs = 1:10)
DF.name = colnames(S8@meta.data)[grepl("DF.classification", colnames(S8@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S8, group.by = "ident") + NoAxes(),
    DimPlot(S8, group.by = DF.name) + NoAxes())
VlnPlot(S8, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S8_singlets = S8[, S8@meta.data[, DF.name] == "Singlet"]
DimPlot(S8_singlets, reduction = "umap")
VlnPlot(S8_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S8_singlets, "S8_singlets.rds")
```

```{r}
sweep.res.S9 <- paramSweep_v3(S9) 
sweep.res.S9 <- summarizeSweep(sweep.res.S9, GT = FALSE) 
bcmvn.S9 <- find.pK(sweep.res.S9)
nExp <- round(ncol(S9) * 0.05)  # expected doublets
S9 <- doubletFinder_v3(S9, pN = 0.25, pK = 0.3, nExp = nExp, PCs = 1:10)
DF.name = colnames(S9@meta.data)[grepl("DF.classification", colnames(S9@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S9, group.by = "ident") + NoAxes(),
    DimPlot(S9, group.by = DF.name) + NoAxes())
VlnPlot(S9, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S9_singlets = S9[, S9@meta.data[, DF.name] == "Singlet"]
DimPlot(S9_singlets, reduction = "umap")
VlnPlot(S9_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S9_singlets, "S9_singlets.rds")
```

```{r}
sweep.res.S10 <- paramSweep_v3(S10) 
sweep.res.S10 <- summarizeSweep(sweep.res.S10, GT = FALSE) 
bcmvn.S10 <- find.pK(sweep.res.S10)
nExp <- round(ncol(S10) * 0.05)  # expected doublets
S10 <- doubletFinder_v3(S10, pN = 0.25, pK = 0.07, nExp = nExp, PCs = 1:10)
DF.name = colnames(S10@meta.data)[grepl("DF.classification", colnames(S10@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S10, group.by = "ident") + NoAxes(),
    DimPlot(S10, group.by = DF.name) + NoAxes())
VlnPlot(S10, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S10_singlets = S10[, S10@meta.data[, DF.name] == "Singlet"]
DimPlot(S10_singlets, reduction = "umap")
VlnPlot(S10_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S10_singlets, "S10_singlets.rds")
```

```{r}
sweep.res.S11 <- paramSweep_v3(S11) 
sweep.res.S11 <- summarizeSweep(sweep.res.S11, GT = FALSE) 
bcmvn.S11 <- find.pK(sweep.res.S11)
nExp <- round(ncol(S11) * 0.05)  # expected doublets
S11 <- doubletFinder_v3(S11, pN = 0.25, pK = 0.3, nExp = nExp, PCs = 1:10)
DF.name = colnames(S11@meta.data)[grepl("DF.classification", colnames(S11@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S11, group.by = "ident") + NoAxes(),
    DimPlot(S11, group.by = DF.name) + NoAxes())
VlnPlot(S11, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S11_singlets = S11[, S11@meta.data[, DF.name] == "Singlet"]
DimPlot(S11_singlets, reduction = "umap")
VlnPlot(S11_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S11_singlets, "S11_singlets.rds")
```

```{r}
sweep.res.S12 <- paramSweep_v3(S12) 
sweep.res.S12 <- summarizeSweep(sweep.res.S12, GT = FALSE) 
bcmvn.S12 <- find.pK(sweep.res.S12)
nExp <- round(ncol(S12) * 0.05)  # expected doublets
S12 <- doubletFinder_v3(S12, pN = 0.25, pK = 0.11, nExp = nExp, PCs = 1:10)
DF.name = colnames(S12@meta.data)[grepl("DF.classification", colnames(S12@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S12, group.by = "ident") + NoAxes(),
    DimPlot(S12, group.by = DF.name) + NoAxes())
VlnPlot(S12, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S12_singlets = S12[, S12@meta.data[, DF.name] == "Singlet"]
DimPlot(S12_singlets, reduction = "umap")
VlnPlot(S12_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S12_singlets, "S12_singlets.rds")
```

```{r}
sweep.res.S13 <- paramSweep_v3(S13) 
sweep.res.S13 <- summarizeSweep(sweep.res.S13, GT = FALSE) 
bcmvn.S13 <- find.pK(sweep.res.S13)
nExp <- round(ncol(S13) * 0.05)  # expected doublets
S13 <- doubletFinder_v3(S13, pN = 0.25, pK = 0.3, nExp = nExp, PCs = 1:10)
DF.name = colnames(S13@meta.data)[grepl("DF.classification", colnames(S13@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S13, group.by = "ident") + NoAxes(),
    DimPlot(S13, group.by = DF.name) + NoAxes())
VlnPlot(S13, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S13_singlets = S13[, S13@meta.data[, DF.name] == "Singlet"]
DimPlot(S13_singlets, reduction = "umap")
VlnPlot(S13_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S13_singlets, "S13_singlets.rds")
```

```{r}
sweep.res.S14 <- paramSweep_v3(S14) 
sweep.res.S14 <- summarizeSweep(sweep.res.S14, GT = FALSE) 
bcmvn.S14 <- find.pK(sweep.res.S14)
nExp <- round(ncol(S14) * 0.05)  # expected doublets
S14 <- doubletFinder_v3(S14, pN = 0.25, pK = 0.05, nExp = nExp, PCs = 1:10)
DF.name = colnames(S14@meta.data)[grepl("DF.classification", colnames(S14@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S14, group.by = "ident") + NoAxes(),
    DimPlot(S14, group.by = DF.name) + NoAxes())
VlnPlot(S14, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S14_singlets = S14[, S14@meta.data[, DF.name] == "Singlet"]
DimPlot(S14_singlets, reduction = "umap")
VlnPlot(S14_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S14_singlets, "S14_singlets.rds")
```

```{r}
sweep.res.S15 <- paramSweep_v3(S15) 
sweep.res.S15 <- summarizeSweep(sweep.res.S15, GT = FALSE) 
bcmvn.S15 <- find.pK(sweep.res.S15)
nExp <- round(ncol(S15) * 0.05)  # expected doublets
S15 <- doubletFinder_v3(S15, pN = 0.25, pK = 0.3, nExp = nExp, PCs = 1:10)
DF.name = colnames(S15@meta.data)[grepl("DF.classification", colnames(S15@meta.data))]
cowplot::plot_grid(ncol = 2, DimPlot(S15, group.by = "ident") + NoAxes(),
    DimPlot(S15, group.by = DF.name) + NoAxes())
VlnPlot(S15, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)
S15_singlets = S15[, S15@meta.data[, DF.name] == "Singlet"]
DimPlot(S15_singlets, reduction = "umap")
VlnPlot(S15_singlets, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), assay = "RNA")
#saveRDS(S15_singlets, "S15_singlets.rds")
```


# Normalization and dimentionality reduction
```{r}
Time.course.No.Day.0 <- merge(S2, c(S3, S4, S5, S6, S7, S8, S9, S10, S11, S12, S13, S15))
```

```{r}
VlnPlot(Time.course.No.Day.0, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(Time.course.No.Day.0, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Time.course.No.Day.0, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

Time.course.No.Day.0 <- NormalizeData(Time.course.No.Day.0)
Time.course.No.Day.0 <- FindVariableFeatures(Time.course.No.Day.0, selection.method = "vst", nfeatures = 3000)
top10 <- head(VariableFeatures(Time.course.No.Day.0), 10)
plot3 <- VariableFeaturePlot(Time.course.No.Day.0)
plot4 <- LabelPoints(plot = plot3, points = top10, repel = TRUE)

Time.course.No.Day.0 <- ScaleData(Time.course.No.Day.0, features = rownames(Time.course.No.Day.0))
Time.course.No.Day.0 <- RunPCA(Time.course.No.Day.0, features = VariableFeatures(object = Time.course.No.Day.0))
VizDimLoadings(Time.course.No.Day.0, dims = 1:2, reduction = "pca")
DimPlot(Time.course.No.Day.0, reduction = "pca")
```

Cell cycle is regressed out using the cell cycle gene lists from the Regev lab (https://github.com/scverse/scanpy_usage/blob/master/180209_cell_cycle/data/regev_lab_cell_cycle_genes.txt) 
```{r}
mouse.s.genes<-c("Mcm5" ,"Pcna", "Tyms", "Fen1", "Mcm2", "Mcm4", "Rrm1", "Ung", "Gins2", "Mcm6", "Cdca7", "Dtl", "Prim1", "Uhrf1", "Mlf1ip", "Hells", "Rfc2", "Rpa2", "Nasp", "Rad51ap1", "Gmnn", "Wdr76", "Slbp", "Ccne2", "Ubr7", "Pold3", "Msh2", "Atad2", "Rad51", "Rrm2", "Cdc45", "Cdc6", "Exo1", "Tipin", "Dscc1", "Blm", "Casp8ap2", "Usp1", "Clspn", "Pola1", "Chaf1b", "Brip1", "E2f8")

mouse.g2m.genes<-c("Hmgb2", "Cdk1", "Nusap1", "Ube2c", "Birc5", "Tpx2", "Top2a", "Ndc80", "Cks2", "Nuf2", "Cks1b", "Mki67", "Tmpo", "Cenpf", "Tacc3", "Fam64a", "Smc4", "Ccnb2", "Ckap2l", "Ckap2", "Aurkb", "Bub1", "Kif11", "Anp32e", "Tubb4b", "Gtse1", "Kif20b", "Hjurp", "Cdca3", "Hn1", "Cdc20", "Ttk", "Cdc25c", "Kif2c", "Rangap1", "Ncapd2", "Dlgap5", "Cdca2", "Cdca8", "Ect2", "Kif23", "Hmmr", "Aurka", "Psrc1", "Anln", "Lbr", "Ckap5", "Cenpe", "Ctcf", "Nek2", "G2e3", "Gas2l3", "Cbx5", "Cenpa")

Time.course.No.Day.0 <- CellCycleScoring(Time.course.No.Day.0, s.features = mouse.s.genes, g2m.features = mouse.g2m.genes, set.ident = TRUE)
RidgePlot(Time.course.No.Day.0, features = c("Pcna", "Top2a", "Mcm6", "Mki67"), ncol = 2)

Time.course.No.Day.0 <- ScaleData(Time.course.No.Day.0, vars.to.regress = c("S.Score", "G2M.Score", "percent.mt"), features = rownames(Time.course.No.Day.0))
```

Principal component analysis
```{r}
Time.course.No.Day.0 <- RunPCA(Time.course.No.Day.0, features = VariableFeatures(object = Time.course.No.Day.0),  npcs = 100, verbose=FALSE)
```

```{r}
ElbowPlot(Time.course.No.Day.0, ndims = 100)
```

Clustering and umap
```{r}
# Determine percent of variation associated with each PC
  pct <- Time.course.No.Day.0@reductions$pca@stdev / sum(Time.course.No.Day.0@reductions$pca@stdev) * 100
  # Determine the difference between variation of PC and subsequent PC
  co2 <- sort(which((pct[1:length(pct)-1] - pct[2:length(pct)]) > 0.1),  decreasing = T)[1] + 1 # last point    where change of % of variation is more than 0.1%.

Time.course.No.Day.0 <- FindNeighbors(Time.course.No.Day.0, dims = 1:7)
Time.course.No.Day.0 <- FindClusters(Time.course.No.Day.0, resolution = 0.6)
Time.course.No.Day.0 <- RunUMAP(Time.course.No.Day.0, dims = 1:7)
# saveRDS(Time.course.No.Day.0, "Time.course.No.Day.0.Clustered.RDS")
```

# Differential expression and cell annotation
DE analysis is performed on the seurat object generated above. The RDS file `Time.course.No.Day.0.Clustered.RDS`  is also available on ArraExpress with accession number E-MTAB-14945. 
```{r}
Time.course.No.Day.0<-readRDS("Time.course.No.Day.0.Clustered.RDS")

Time.Course.Cluster.0.markers <- FindMarkers(Time.course.No.Day.0, ident.1 = "0", assay = "RNA", logfc.threshold = 0.585, min.pct = 0.25, only.pos = T) 
Time.Course.Cluster.0.markers <- Time.Course.Cluster.0.markers %>%
filter(p_val_adj < 0.05)
Time.Course.Cluster.1.markers <- FindMarkers(Time.course.No.Day.0, ident.1 = "1", assay = "RNA", logfc.threshold = 0.585, min.pct = 0.25, only.pos = T)
Time.Course.Cluster.1.markers <- Time.Course.Cluster.1.markers %>%
filter(p_val_adj < 0.05)
Time.Course.Cluster.2.markers <- FindMarkers(Time.course.No.Day.0, ident.1 = "2", assay = "RNA", logfc.threshold = 0.585, min.pct = 0.25, only.pos = T)
Time.Course.Cluster.2.markers <- Time.Course.Cluster.2.markers %>%
filter(p_val_adj < 0.05)
Time.Course.Cluster.3.markers <- FindMarkers(Time.course.No.Day.0, ident.1 = "3", assay = "RNA", logfc.threshold = 0.585, min.pct = 0.25, only.pos = T)
Time.Course.Cluster.3.markers <- Time.Course.Cluster.3.markers %>%
filter(p_val_adj < 0.05)
Time.Course.Cluster.4.markers <- FindMarkers(Time.course.No.Day.0, ident.1 = "4", assay = "RNA", logfc.threshold = 0.585, min.pct = 0.25, only.pos = T)
Time.Course.Cluster.4.markers <- Time.Course.Cluster.4.markers %>%
filter(p_val_adj < 0.05)
Time.Course.Cluster.5.markers <- FindMarkers(Time.course.No.Day.0, ident.1 = "5", assay = "RNA", logfc.threshold = 0.585, min.pct = 0.25, only.pos = T)
Time.Course.Cluster.5.markers <- Time.Course.Cluster.5.markers %>%
filter(p_val_adj < 0.05)
Time.Course.Cluster.6.markers <- FindMarkers(Time.course.No.Day.0, ident.1 = "6", assay = "RNA", logfc.threshold = 0.585, min.pct = 0.25, only.pos = T)
Time.Course.Cluster.6.markers <- Time.Course.Cluster.6.markers %>%
filter(p_val_adj < 0.05)
```

```{r}
Time.course.No.Day.0@active.ident <- factor(x = Time.course.No.Day.0@active.ident, levels = c('0', '3', '1', '2', '5', '4', '6'))
```

# UMAP plots and Z-scores
#Figure 4A
```{r}
pA <- DimPlot(Time.course.No.Day.0, reduction = "umap", label = T, pt.size = 1) + NoLegend()
```

#Figure 4B
```{r}
pB <- DimPlot(Time.course.No.Day.0, reduction = "umap", label = F, pt.size = 1, group.by = "Time.point") + labs(title = "")
```

```{r}
Time.course.No.Day.0@meta.data <- Time.course.No.Day.0@meta.data %>%
  mutate(Annotation = case_when(
    RNA_snn_res.0.6 == "0" ~ "Day 1",
    RNA_snn_res.0.6 == "1" ~ "Early iCAFs",
    RNA_snn_res.0.6 == "2" ~ "iCAFs/ifCAFs",
    RNA_snn_res.0.6 == "3" ~ "Days 2/3",
    RNA_snn_res.0.6 == "4" ~ "myCAFs",
    RNA_snn_res.0.6 == "5" ~ "Late iCAFs",
    RNA_snn_res.0.6 == "6" ~ "ifCAFs"
  ))
```

```{r}
# saveRDS(Time.course.No.Day.0, "Time.course.No.Day.0.Clustered.Annotated.RDS")
# write.table(Time.course.No.Day.0@meta.data, "Time.course.No.Day.0.Meta.Data.txt", sep = "\t")
```


# Figure 4C
```{r}
Time.course.No.Day.0<-readRDS("Time.course.No.Day.0.Clustered.Annotated.RDS")

Time.point.cluster.prop <-  Time.course.No.Day.0@meta.data %>% 
  group_by(RNA_snn_res.0.6, Time.point) %>% 
  summarise(count=n()) %>% 
  group_by(RNA_snn_res.0.6) %>% 
  mutate(prop=count/sum(count)) %>% 
  ungroup()

pC <- ggplot(Time.point.cluster.prop, aes(x=factor(RNA_snn_res.0.6, levels = c('0', '3', '1', '2', '5', '4', '6')), y=prop, fill=Time.point)) +
  geom_bar(stat="identity", position="stack") +
  theme_bw() + 
  theme(panel.border = element_rect(colour = "black", fill = NA, size= 0.5),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank()) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 1)) +
  xlab("Cluster") +
  ylab("Proportion") + 
  scale_fill_discrete(name = "Time point")
```



# Signature Z-scores
Signature Z-scores from day 6 cocultures. 
A function that adds a z-score in the metadata of a seurat object. The signature should be a character vector
```{r}
Add_Z<-function(in_SO, in_sig, Z_name){
  
  DefaultAssay(in_SO) <- "RNA"
  #verifies the genes are part of the seurat object
  in_sig<-in_sig[in_sig %in% rownames(in_SO)]
  
   # Genes which are 0 in >95% of cells are removed
  long_NC <- in_SO@assays$RNA@data[in_sig,] %>%
    round(2) %>% as.matrix() %>% as.data.frame() %>%
    rownames_to_column("Gene") %>%
    gather(key = "Cell", value = "NC",-Gene) %>% 
    group_by(Gene) %>% 
    summarise(prop_zeros = sum(NC == 0)/n()) %>% 
    filter(prop_zeros <= 0.95) %>% 
    dplyr::select(-prop_zeros)
    
  # NC is extracted from the SO
    long_NC <- in_SO@assays$RNA@data[long_NC$Gene,] %>%
    round(2) %>% as.matrix() %>% as.data.frame() %>%
    rownames_to_column("Gene") %>%
    gather(key = "Cell", value = "NC",-Gene) 
  
  # Signature scores are calculated for each cell
  get_zs<-function(in_sign){
    Z_scores<-long_NC %>%
      filter(Gene %in% in_sign) %>% 
      group_by(Gene) %>%
      mutate(Z = (NC - mean(NC)) / sd(NC)) %>%
      drop_na() %>% 
      group_by(Cell) %>%
      summarise(sign_Z = mean(Z) %>% round(2)) %>%
      dplyr::rename(!!Z_name := sign_Z)
    
    return(Z_scores)
  }
  
  Z_df<-get_zs(in_sig) 
 
  # Z-score is added to the metadata
  in_SO@meta.data<-in_SO@meta.data %>% 
    rownames_to_column("Cell") %>% 
    left_join(Z_df, by="Cell") %>% 
    column_to_rownames("Cell")
  
  return(in_SO)
}
```

The markers lists are retrieved from the analysis described in R markdown `Figure_3_S6_data_analysis.rmd`
```{r}
get_RNA_markers<-function(SO, id1, logfc_cutoff=0.585, min_pct_cutoff=0.25, padj_cutoff=0.05){
  markers_df<-FindMarkers(SO, ident.1 = id1, assay = "RNA", logfc.threshold = logfc_cutoff, min.pct = min_pct_cutoff, only.pos = T)
  markers_df<-markers_df %>% filter(p_val_adj<padj_cutoff)
  return(markers_df)
}

PSC.cocul.Cluster.0.markers<-get_RNA_markers(SO=PSC.singlets.rPCA.integrated, id1 = "0")
PSC.cocul.Cluster.1.markers<-get_RNA_markers(SO=PSC.singlets.rPCA.integrated, id1 = "1")
PSC.cocul.Cluster.2.markers<-get_RNA_markers(SO=PSC.singlets.rPCA.integrated, id1 = "2")
PSC.cocul.Cluster.3.markers<-get_RNA_markers(SO=PSC.singlets.rPCA.integrated, id1 = "3")
PSC.cocul.Cluster.4.markers<-get_RNA_markers(SO=PSC.singlets.rPCA.integrated, id1 = "4")
PSC.cocul.Cluster.5.markers<-get_RNA_markers(SO=PSC.singlets.rPCA.integrated, id1 = "5")
```

Signature Z-scores are added to the metadata. Note that this must be done only once.
```{r}
Time.course.No.Day.0 <- Add_Z(in_SO = Time.course.No.Day.0, in_sig = rownames(PSC.cocul.Cluster.0.markers), Z_name = "Cluster 0 Co-culture Z-Score")
Time.course.No.Day.0 <- Add_Z(in_SO = Time.course.No.Day.0, in_sig = rownames(PSC.cocul.Cluster.1.markers), Z_name = "Cluster 1 Co-culture Z-Score")
Time.course.No.Day.0 <- Add_Z(in_SO = Time.course.No.Day.0, in_sig = rownames(PSC.cocul.Cluster.2.markers), Z_name = "Cluster 2 Co-culture Z-Score")
Time.course.No.Day.0 <- Add_Z(in_SO = Time.course.No.Day.0, in_sig = rownames(PSC.cocul.Cluster.3.markers), Z_name = "Cluster 3 Co-culture Z-Score")
Time.course.No.Day.0 <- Add_Z(in_SO = Time.course.No.Day.0, in_sig = rownames(PSC.cocul.Cluster.4.markers), Z_name = "Cluster 4 Co-culture Z-Score")
Time.course.No.Day.0 <- Add_Z(in_SO = Time.course.No.Day.0, in_sig = rownames(PSC.cocul.Cluster.5.markers), Z_name = "Cluster 5 Co-culture Z-Score")

# saveRDS(Time.course.No.Day.0, "Time.course.No.Day.0.Clustered.Annotated.Zscore.RDS")
```

#4D-F
```{r}
pD.1 <- VlnPlot(Time.course.No.Day.0, features = "Cluster 0 Co-culture Z-Score", pt.size = 0) + NoLegend() + labs(title = "Co-culture iCAF Z-score") & geom_hline(yintercept = 0) 
pE.1 <- VlnPlot(Time.course.No.Day.0, features = "Cluster 1 Co-culture Z-Score", pt.size = 0) + NoLegend() + labs(title = "Co-culture myCAF Z-score") & geom_hline(yintercept = 0) 
pF.1 <- VlnPlot(Time.course.No.Day.0, features = "Cluster 4 Co-culture Z-Score", pt.size = 0, y.max = 1.2) + NoLegend() + labs(title = "Co-culture ifCAF Z-score") & geom_hline(yintercept = 0) 

pD.2 <- FeaturePlot(Time.course.No.Day.0, features = "Cluster 0 Co-culture Z-Score", cols = c("lightblue", "red")) + ggtitle(label = "")
pE.2 <- FeaturePlot(Time.course.No.Day.0, features = "Cluster 1 Co-culture Z-Score", cols = c("lightblue", "red")) + ggtitle(label = "")
pF.2 <- FeaturePlot(Time.course.No.Day.0, features = "Cluster 4 Co-culture Z-Score", cols = c("lightblue", "red"), min.cutoff = -0.5, max.cutoff = 1) + ggtitle(label = "")

```

#Save Figures 4A-F
```{r}
Fig4_AtoF_plot <- plot_grid(pA, pB, pC, pD.1, pE.1, pF.1, pD.2, pE.2, pF.2, nrow = 3, ncol = 3)

if (!dir.exists("Figure_4")) {dir.create("Figure_4")}
ggsave(plot = Fig4_AtoF_plot, "Figure_4A_F.pdf", width = 15, height = 4.4444444*1.5, units = "in", path = "Figure_4/")
```


# Trajectory predicitions
SlingShot is used to predict trajectories
The objects are saved as separate matrices for input in Slingshot
```{r}
Time.course.No.Day.0<-readRDS("Time.course.No.Day.0.Clustered.Annotated.Zscore.RDS")


dimred <- Time.course.No.Day.0@reductions$umap@cell.embeddings
clustering <- Time.course.No.Day.0$RNA_snn_res.0.6
counts <- as.matrix(Time.course.No.Day.0@assays$RNA@counts[Time.course.No.Day.0@assays$RNA@var.features, ])
```

The default Slingshot lineage identification is performed
```{r}
set.seed(1)
lineages <- getLineages(data = dimred,
                        clusterLabels = clustering,
                        start.clus = "0", end.clus = c("4","5","6")) #define where to start the trajectories

lineages
```

```{r}
# Plot the lineages
pal <- c(RColorBrewer::brewer.pal(9, "Set1"), RColorBrewer::brewer.pal(8, "Set2"))
seur_pal <- c("#F8766D", "#7CAE00", "#00C19A", "#CD9600", "#8494FF", "#00B8E7",  "#FF61CC")
par(mfrow = c(1, 2))
plot(dimred[, 1:2], col = pal[clustering], cex = 0.5, pch = 16)
for (i in levels(clustering)) {
    text(mean(dimred[clustering == i, 1]), mean(dimred[clustering == i, 2]), labels = i, font = 2)
}
plot(dimred[, 1:2], col = seur_pal[clustering], cex = 0.5, pch = 16) +
lines(SlingshotDataSet(lineages), lwd = 3, col = "black")
```

# Figure 4G
```{r}
seur_pal <- c("#F8766D", "#7CAE00", "#00C19A", "#CD9600", "#8494FF", "#00B8E7", "#FF61CC")
Slingshot_plot<-plot(dimred[, 1:2], col = seur_pal[clustering], cex = 0.5, pch = 16) +
lines(SlingshotDataSet(lineages), lwd = 3, col = "black")


```

```{r}
Fig_4G_plot <- ggplotify::as.ggplot(function() plot(dimred[, 1:2], col = seur_pal[clustering], cex = 0.5, pch = 16) +
lines(SlingshotDataSet(lineages), lwd = 3, col = "black"))

if (!dir.exists("Figure_4")) {dir.create("Figure_4")}
ggsave(plot = Fig_4G_plot, "Figure_4G_SlingShot_Trajectory.pdf", width = 3.25, height = 4.4444444, units = "in", path = "Figure_4/")
```

# Figure 4H
The pseudotime values are extracted
```{r}
crv1 <- getCurves(lineages)

plot(dimred, col = pal[clustering], asp = 1, pch = 16)+
lines(SlingshotDataSet(crv1), lwd = 3, col = 'black')
```

```{r}
sds <- slingshot(Embeddings(Time.course.No.Day.0, "umap"), clusterLabels = Time.course.No.Day.0@meta.data$RNA_snn_res.0.6, 
                 start.clus = "0", end.clus = c("4","5","6"))

pal <- viridis(100, end = 0.95)
Time.course.pseudotime <- slingPseudotime(sds)

ifCAf_pt_plot<-ggplotify::as.ggplot(function() plot(reducedDim(SlingshotDataSet(sds)), col = pal[cut(Time.course.pseudotime[,"Lineage1"], breaks = 100)], pch = 16, cex = 0.5, main = "Lineage1")+
  lines(SlingshotDataSet(sds), lwd = 2, col = 'black', type = 'lineages'))
iCAf_pt_plot<-ggplotify::as.ggplot(function() plot(reducedDim(SlingshotDataSet(sds)), col = pal[cut(Time.course.pseudotime[,"Lineage2"], breaks = 100)], pch = 16, cex = 0.5, main = "Lineage2")+
  lines(SlingshotDataSet(sds), lwd = 2, col = 'black', type = 'lineages'))
myCAf_pt_plot<-ggplotify::as.ggplot(function() plot(reducedDim(SlingshotDataSet(sds)), col = pal[cut(Time.course.pseudotime[,"Lineage3"], breaks = 100)], pch = 16, cex = 0.5, main = "Lineage3")+
  lines(SlingshotDataSet(sds), lwd = 2, col = 'black', type = 'lineages'))
```

```{r}
Pseudo.lineage<- as.data.frame(Time.course.pseudotime)

Time.course.No.Day.0.meta.data <- Time.course.No.Day.0@meta.data %>%
  rownames_to_column("Cell")

Pseudo.lineage.1 <- Pseudo.lineage %>%
  dplyr::select(Lineage1)%>%
  drop_na() %>%
  rownames_to_column("Cell") %>% 
  left_join(Time.course.No.Day.0.meta.data, by = "Cell") %>%
  dplyr::select(c(Time.point, Lineage1, RNA_snn_res.0.6, "Cluster.4.Score"="Cluster 4 Co-culture Z-Score", Cell))

Pseudo.lineage.2 <- Pseudo.lineage %>%
  dplyr::select(Lineage2)%>%
  drop_na() %>%
  rownames_to_column("Cell") %>% 
  left_join(Time.course.No.Day.0.meta.data, by = "Cell") %>%
  dplyr::select(c(Time.point, Lineage2, RNA_snn_res.0.6, "Cluster.0.Score"="Cluster 0 Co-culture Z-Score", Cell))

Pseudo.lineage.3 <- Pseudo.lineage %>%
  dplyr::select(Lineage3)%>%
  drop_na() %>%
  rownames_to_column("Cell") %>% 
  left_join(Time.course.No.Day.0.meta.data, by = "Cell") %>%
  dplyr::select(c(Time.point, Lineage3, RNA_snn_res.0.6, "Cluster.1.Score"="Cluster 1 Co-culture Z-Score", Cell))
```

Plots are made representing day 6 CAF signature Z-Scores as a function of pseudotime
```{r}
p1 <- ggplot(Pseudo.lineage.1, aes(x=Lineage1, y=Time.point, color = Time.point)) +
  geom_jitter() + theme_bw() + xlab("Pseudotime") + ggtitle("Lineage 1") + ylab("") + theme(legend.title = element_text(size = 0)) + theme(plot.title = element_text(hjust = 0.5)) +
  geom_vline(xintercept=16,linetype=3) + NoLegend()

p2 <- ggplot(Pseudo.lineage.1, aes(x=Lineage1, y=Cluster.4.Score, color = RNA_snn_res.0.6)) +
  geom_point() + theme_bw() + xlab("Pseudotime") + ggtitle("") + ylab("ifCAF Z-Score") + theme(legend.title = element_text(size = 0)) + geom_hline(yintercept = 0) + ylim(-0.75, 1.5) + geom_vline(xintercept=16,linetype=3) +
  scale_color_manual(values=c("0"="#F8766D", "1"="#7CAE00", "2" = "#00C19A", "3" = "#CD9600", "4" = "#8494FF", "5" = "#00B8E7", "6" = "#FF61CC"))

p3 <- ggplot(Pseudo.lineage.2, aes(x=Lineage2, y=Time.point, color = Time.point)) +
  geom_jitter() + theme_bw() + xlab("Pseudotime") + ggtitle("Lineage 2") + ylab("") + theme(legend.title = element_text(size = 0)) + theme(plot.title = element_text(hjust = 0.5)) + NoLegend()

p4 <- ggplot(Pseudo.lineage.2, aes(x=Lineage2, y=Cluster.0.Score, color = RNA_snn_res.0.6)) +
  geom_point() + theme_bw() + xlab("Pseudotime") + ggtitle("") + ylab("iCAF Z-Score") + theme(legend.title = element_text(size = 0))  + geom_hline(yintercept = 0) +
  scale_color_manual(values=c("0"="#F8766D", "1"="#7CAE00", "2" = "#00C19A", "3" = "#CD9600", "4" = "#8494FF", "5" = "#00B8E7", "6" = "#FF61CC"))

p5 <- ggplot(Pseudo.lineage.3, aes(x=Lineage3, y=Time.point, color = Time.point)) +
  geom_jitter() + theme_bw() + xlab("Pseudotime") + ggtitle("Lineage 3") + ylab("") + theme(legend.title = element_text(size = 0)) + theme(plot.title = element_text(hjust = 0.5)) + NoLegend() 

p6 <- ggplot(Pseudo.lineage.3, aes(x=Lineage3, y=Cluster.1.Score, color = RNA_snn_res.0.6)) +
  geom_point() + theme_bw() + xlab("Pseudotime") + ggtitle("") + ylab("myCAF Z-Score") + theme(legend.title = element_text(size = 0)) + geom_hline(yintercept = 0) +
  scale_color_manual(values=c("0"="#F8766D", "1"="#7CAE00", "3" = "#CD9600", "4" = "#8494FF", "5" = "#00B8E7", "6" = "#FF61CC"))
```

#Figure 4I to K
```{r}
Fig_4ItoK_plot <- plot_grid(p1, p3, p5, p2, p4, p6, nrow = 2)

if (!dir.exists("Figure_4")) {dir.create("Figure_4")}
ggsave(plot = Fig_4ItoK_plot, "Figure_4I_to_K.pdf", width = 15, height = 4.4444444, units = "in", path = "Figure_4/")
```


# Transcription factor activities
Transcription factor activities are predicted with DecoupleR
```{r}
library(decoupleR)
```


```{r}
net <- get_collectri(organism = "mouse", split_complexes = FALSE)

Time.course.No.Day.0<-readRDS("Time.course.No.Day.0.Clustered.Annotated.Zscore.RDS")

mat<- as.matrix(Time.course.No.Day.0@assays$RNA@data)
acts_ulm_time_course <- run_ulm(mat=mat, net=net, .source='source', .target='target',
                .mor='mor', minsize = 20)
# write.table(acts_ulm_time_course, "../Data/acts_ulm_time_course.txt", sep = "\t")
```

```{r}
Time.course.No.Day.0[['tfsulm']] <- acts_ulm_time_course %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)


# Change assay
DefaultAssay(object = Time.course.No.Day.0) <- "tfsulm"

# Scale the data
TF.Activity.Time.course.No.Day.0 <- ScaleData(Time.course.No.Day.0)

#saveRDS(TF.Activity.Time.course.No.Day.0, "TF.Activity.Time.course.No.Day.0.RDS")
```

Correlations between transcription factor activity and pseudotime are calculated
```{r}
TF.Activity.Time.course.No.Day.0 <- readRDS("TF.activity.Time.course.No.Day.0.RDS")
```

```{r}
Cluster.0.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "0", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.1.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "1", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.2.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "2", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.3.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "3", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.4.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "4", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.5.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "5", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.6.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "6", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.3.vs.0.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "3", ident.2 = "0", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.1.vs.3.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "1", ident.2 = "3", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.2.vs.1.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "2", ident.2 = "1", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.6.vs.2.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "6", ident.2 = "2", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.5.vs.2.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "5", ident.2 = "2", assay = "tfsulm", only.pos = T, slot = "scale.data")
Cluster.4.vs.3.TFs <- FindMarkers(TF.Activity.Time.course.No.Day.0, ident.1 = "4", ident.2 = "3", assay = "tfsulm", only.pos = T, slot = "scale.data")
```

```{r}
TF.Activity.per.cell <- TF.Activity.Time.course.No.Day.0@assays$tfsulm@scale.data
TF.Activity.per.cell <- TF.Activity.per.cell %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Cell")
```

The TF activities are extracted from object as a long dataframe
```{r}
TF.Activity.Time.course.No.Day.0@active.ident <- Time.course.No.Day.0@active.ident
TF.Activity.df <- t(as.matrix(TF.Activity.Time.course.No.Day.0@assays$tfsulm@scale.data)) %>%
  as.data.frame() %>%
  mutate(cluster = TF.Activity.Time.course.No.Day.0@active.ident) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))
```

```{r}
Pseudo.lineage.1.subset <- Pseudo.lineage.1 %>%
  dplyr::filter(Lineage1 <= 16)
Pseudo.lineage1.pseudotime.vs.TF.Activity <- Pseudo.lineage.1.subset %>%
  left_join(TF.Activity.per.cell, by = "Cell") %>%
  dplyr::select(-c(Cluster.4.Score, Time.point, RNA_snn_res.0.6)) %>%
  column_to_rownames("Cell")

Pseudo.lineage.1.pseudotime.vs.TF.Activity.Cor.df <- data.frame(var1 = character(),
                     var2 = character(),
                     cor_coef = numeric(),
                     p_value = numeric(),
                     stringsAsFactors = FALSE)

for (i in names(Pseudo.lineage1.pseudotime.vs.TF.Activity)){
  if (i != "Lineage1"){
    cor_test <- cor.test(Pseudo.lineage1.pseudotime.vs.TF.Activity$Lineage1, Pseudo.lineage1.pseudotime.vs.TF.Activity[,i])
    Pseudo.lineage.1.pseudotime.vs.TF.Activity.Cor.df <- rbind(Pseudo.lineage.1.pseudotime.vs.TF.Activity.Cor.df, c(i, "Lineage1", cor_test$estimate, cor_test$p.value))
  }
}

colnames(Pseudo.lineage.1.pseudotime.vs.TF.Activity.Cor.df) <- c("TF", "Pseudotime", "Cor_coef", "P-value")

Pseudo.lineage.1.pseudotime.vs.TF.Activity.Cor.df <- as.data.frame(Pseudo.lineage.1.pseudotime.vs.TF.Activity.Cor.df)

top_TFs_lin_1 <- Pseudo.lineage.1.pseudotime.vs.TF.Activity.Cor.df %>%
  mutate_at(vars(Cor_coef), as.numeric) %>%
  top_n(5, abs(Cor_coef))
```

```{r}
Pseudo.lineage.2.pseudotime.vs.TF.Activity <- Pseudo.lineage.2 %>%
  left_join(TF.Activity.per.cell, by = "Cell") %>%
  dplyr::select(-c(Cluster.0.Score, Time.point, RNA_snn_res.0.6)) %>%
  column_to_rownames("Cell")

Pseudo.lineage.2.pseudotime.vs.TF.Activity.Cor.df <- data.frame(var1 = character(),
                     var2 = character(),
                     cor_coef = numeric(),
                     p_value = numeric(),
                     stringsAsFactors = FALSE)

for (i in names(Pseudo.lineage.2.pseudotime.vs.TF.Activity)){
  if (i != "Lineage2"){
    cor_test <- cor.test(Pseudo.lineage.2.pseudotime.vs.TF.Activity$Lineage2, Pseudo.lineage.2.pseudotime.vs.TF.Activity[,i])
    Pseudo.lineage.2.pseudotime.vs.TF.Activity.Cor.df <- rbind(Pseudo.lineage.2.pseudotime.vs.TF.Activity.Cor.df, c(i, "Lineage2", cor_test$estimate, cor_test$p.value))
  }
}

colnames(Pseudo.lineage.2.pseudotime.vs.TF.Activity.Cor.df) <- c("TF", "Lineage2", "Cor_coef", "P-value")

Pseudo.lineage.2.pseudotime.vs.TF.Activity.Cor.df <- as.data.frame(Pseudo.lineage.2.pseudotime.vs.TF.Activity.Cor.df)

top_TFs_lin_2 <- Pseudo.lineage.2.pseudotime.vs.TF.Activity.Cor.df %>%
  mutate_at(vars(Cor_coef), as.numeric) %>%
  top_n(5, abs(Cor_coef))
```


```{r}
Pseudo.lineage.3.pseudotime.vs.TF.Activity <- Pseudo.lineage.3 %>%
  left_join(TF.Activity.per.cell, by = "Cell") %>%
  dplyr::select(-c(Cluster.1.Score, Time.point, RNA_snn_res.0.6)) %>%
  column_to_rownames("Cell")

Pseudo.lineage.3.pseudotime.vs.TF.Activity.Cor.df <- data.frame(var1 = character(),
                     var2 = character(),
                     cor_coef = numeric(),
                     p_value = numeric(),
                     stringsAsFactors = FALSE)

for (i in names(Pseudo.lineage.3.pseudotime.vs.TF.Activity)){
  if (i != "Lineage3"){
    cor_test <- cor.test(Pseudo.lineage.3.pseudotime.vs.TF.Activity$Lineage3, Pseudo.lineage.3.pseudotime.vs.TF.Activity[,i])
    Pseudo.lineage.3.pseudotime.vs.TF.Activity.Cor.df <- rbind(Pseudo.lineage.3.pseudotime.vs.TF.Activity.Cor.df, c(i, "Lineage3", cor_test$estimate, cor_test$p.value))
  }
}

colnames(Pseudo.lineage.3.pseudotime.vs.TF.Activity.Cor.df) <- c("TF", "Lineage3", "Cor_coef", "P-value")

Pseudo.lineage.3.pseudotime.vs.TF.Activity.Cor.df <- as.data.frame(Pseudo.lineage.3.pseudotime.vs.TF.Activity.Cor.df)

top_TF1_lin_3 <- Pseudo.lineage.3.pseudotime.vs.TF.Activity.Cor.df %>%
  mutate_at(vars(Cor_coef), as.numeric) %>%
  top_n(1, abs(Cor_coef))
```

# Figure 4L
```{r}
p1 <- ggplot(Pseudo.lineage1.pseudotime.vs.TF.Activity, aes(Lineage1, Stat2)) + 
  geom_point() + stat_smooth(method = "lm") +
  background_grid(major = 'y', minor = "none") +
  panel_border() +
  stat_cor(aes(label = ..r.label..),method = "spearman", label.x = -0.5, label.y = 5, size = 4)  +
  ylab("Stat2 Activity Score") + 
  theme_bw() + ggtitle("Lineage 1") + theme(plot.title = element_text(hjust = 0.5)) +
  geom_vline(xintercept=16,linetype=3)

p2 <- ggplot(Pseudo.lineage.2.pseudotime.vs.TF.Activity, aes(Lineage2, Hif1a)) + 
  geom_point() + stat_smooth(method = "lm") +
  background_grid(major = 'y', minor = "none") +
  panel_border() +
  stat_cor(aes(label = ..r.label..),method = "spearman", label.x = -0.5, label.y = 5, size = 4)  +
  ylab("Hif1a Activity Score") + 
  theme_bw() + ggtitle("Lineage 2") + theme(plot.title = element_text(hjust = 0.5))

p3 <- ggplot(Pseudo.lineage.3.pseudotime.vs.TF.Activity, aes(Lineage3, Smad3)) + 
  geom_point() + stat_smooth(method = "lm") +
  background_grid(major = 'y', minor = "none") +
  panel_border() +
  stat_cor(aes(label = ..r.label..),method = "spearman", label.x = -0.5, label.y = 5, size = 4)  +
  ylab("Smad3 Activity Score") + 
  theme_bw() + ggtitle("Lineage 3") + theme(plot.title = element_text(hjust = 0.5))

Fig_4L_plot <- plot_grid(p1, p2, p3, nrow = 3)

if (!dir.exists("Figure_4")) {dir.create("Figure_4")}
ggsave(plot = Fig_4L_plot, "Figure_4L_Top_TF_Correlations.pdf", width = 3.75, height = 4.4444444, units = "in", path = "Figure_4/")
```

# Figure 4M
correlation coefficient for selected TFs
```{r}
Lineage1_pseudotime_TF_cor <- Pseudo.lineage.1.pseudotime.vs.TF.Activity.Cor.df %>%
  mutate_at(vars(Cor_coef), as.numeric) %>%
  select(c(TF, Cor_coef)) %>%
  rename("Cor_coef" = "Cor_coef_Lin1")

Lineage2_pseudotime_TF_cor <- Pseudo.lineage.2.pseudotime.vs.TF.Activity.Cor.df %>%
  mutate_at(vars(Cor_coef), as.numeric) %>%
  select(c(TF, Cor_coef)) %>%
  rename("Cor_coef" = "Cor_coef_Lin2")

Lineage3_pseudotime_TF_cor <- Pseudo.lineage.3.pseudotime.vs.TF.Activity.Cor.df %>%
  mutate_at(vars(Cor_coef), as.numeric) %>%
  select(c(TF, Cor_coef)) %>%
  rename("Cor_coef" = "Cor_coef_Lin3")

Lineages_pseudotime_TF_cor <- Lineage1_pseudotime_TF_cor %>%
  left_join(Lineage2_pseudotime_TF_cor, by = "TF") %>%
  left_join(Lineage3_pseudotime_TF_cor, by = "TF") %>%
  mutate(Lin_1_Cor_Coef_Dif = (Cor_coef_Lin1 - ((Cor_coef_Lin2 + Cor_coef_Lin3)/2)) + Cor_coef_Lin1) %>%
  mutate(Lin_2_Cor_Coef_Dif = (Cor_coef_Lin2 - ((Cor_coef_Lin1 + Cor_coef_Lin3)/2)) + Cor_coef_Lin2) %>%
  mutate(Lin_3_Cor_Coef_Dif = (Cor_coef_Lin3 - ((Cor_coef_Lin1 + Cor_coef_Lin2)/2)) + Cor_coef_Lin3)

Top.Lin1.Enriched.TFs <- Lineages_pseudotime_TF_cor %>%
  top_n(10, abs(Lin_1_Cor_Coef_Dif))

Top.Lin2.Enriched.TFs <- Lineages_pseudotime_TF_cor %>%
  top_n(10, abs(Lin_2_Cor_Coef_Dif))

Top.Lin3.Enriched.TFs <- Lineages_pseudotime_TF_cor %>%
  top_n(10, abs(Lin_3_Cor_Coef_Dif))

Selected_TFs_Enriched_All_Lin <- Lineage1_pseudotime_TF_cor %>%
  left_join(Lineage2_pseudotime_TF_cor, by = "TF") %>%
  left_join(Lineage3_pseudotime_TF_cor, by = "TF") %>%
  filter(TF %in% unique(c(Top.Lin1.Enriched.TFs$TF, Top.Lin2.Enriched.TFs$TF, Top.Lin3.Enriched.TFs$TF))) %>%
  column_to_rownames("TF")


Selected_TFs_Enriched_All_Lin <- Selected_TFs_Enriched_All_Lin[match(unique(c(Top.Lin1.Enriched.TFs$TF, Top.Lin2.Enriched.TFs$TF, Top.Lin3.Enriched.TFs$TF)), rownames(Selected_TFs_Enriched_All_Lin)),]

breaksList = seq(-1, 1, by = 0.01)
pheatmap(Selected_TFs_Enriched_All_Lin, breaks = breaksList, cluster_rows = F, cluster_cols = F, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(length(breaksList)))
```

```{r}
#write.table(Lineage1_pseudotime_TF_cor, "Lineage1_pseudotime_TF_cor.txt", sep = "\t")
#write.table(Lineage2_pseudotime_TF_cor, "Lineage2_pseudotime_TF_cor.txt", sep = "\t")
#write.table(Lineage3_pseudotime_TF_cor, "Lineage3_pseudotime_TF_cor.txt", sep = "\t")
```


```{r}
breaksList = seq(-1, 1, by = 0.01)
Fig_4M_plot <- as.ggplot(function() pheatmap(Selected_TFs_Enriched_All_Lin, fontsize_row = 10, breaks = breaksList, cluster_rows = F, cluster_cols = F, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(length(breaksList))))

if (!dir.exists("Figure_4")) {dir.create("Figure_4")}
ggsave(plot = myplot, "Figure_4M_Top_TF_Correlations_heatmap.pdf", width = 11.25, height = 4.4444444, units = "in", path = "Figure_4/")
```


# Correlation of gene expression with pseudotime and GSEA
```{r}
Time.course.No.Day.0<-readRDS("Time.course.No.Day.0.Clustered.Annotated.Zscore.RDS")

Time.course.No.Day.0.Norm.Counts <- as.data.frame(Time.course.No.Day.0@assays$RNA@data) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Cell")

Pseudo.lineage.1.vs.Gene.Expression <- Pseudo.lineage.1.subset %>%
  left_join(Time.course.No.Day.0.Norm.Counts, by = "Cell") %>%
  dplyr::select(-c(Cluster.4.Score, Time.point, RNA_snn_res.0.6)) %>%
  column_to_rownames("Cell")   

#Pseudo.lineage.1.vs.Gene.Expression <- Pseudo.lineage.1.vs.Gene.Expression[, apply(Pseudo.lineage.1.vs.Gene.Expression, 2, function(x) !all(x == 0))]

# compute percentage of 0 values in each column
percent_zero <- apply(Pseudo.lineage.1.vs.Gene.Expression == 0, 2, mean)


# subset columns where less than 75% of values are 0, using the %in% operator
Pseudo.lineage.1.vs.Gene.Expression <- Pseudo.lineage.1.vs.Gene.Expression[, percent_zero < 0.9]




slope_corr <- function(x, y) {
  out <- summary(lm(y~x))
  coef <- out$coefficients[2,1]
  cor <- cor(x,y)
  pval <- out$coefficients[2,4]
  return(c(coef, cor, pval))
}

# Apply the slope_corr function to Column A with all other columns
Pseudo.lineage.1.vs.Gene.Expression.Cor.df <- apply(Pseudo.lineage.1.vs.Gene.Expression[, -1], 2, function(x) slope_corr(Pseudo.lineage.1.vs.Gene.Expression$Lineage1, x))

# Convert results to dataframe
Pseudo.lineage.1.vs.Gene.Expression.Cor.df <- t(data.frame(Pseudo.lineage.1.vs.Gene.Expression.Cor.df))

colnames(Pseudo.lineage.1.vs.Gene.Expression.Cor.df) <- c("Slope", "Correlation", "P_value")

Pseudo.lineage.1.vs.Gene.Expression.Cor.df <- as.data.frame(Pseudo.lineage.1.vs.Gene.Expression.Cor.df)
```

```{r}
Pseudo.lineage.2.vs.Gene.Expression <- Pseudo.lineage.2 %>%
  left_join(Time.course.No.Day.0.Norm.Counts, by = "Cell") %>%
  dplyr::select(-c(Cluster.0.Score, Time.point, RNA_snn_res.0.6)) %>%
  column_to_rownames("Cell")   

#Pseudo.lineage.2.vs.Gene.Expression <- Pseudo.lineage.2.vs.Gene.Expression[, apply(Pseudo.lineage.2.vs.Gene.Expression, 2, function(x) !all(x == 0))]

# compute percentage of 0 values in each column
percent_zero <- apply(Pseudo.lineage.2.vs.Gene.Expression == 0, 2, mean)


# subset columns where less than 75% of values are 0, using the %in% operator
Pseudo.lineage.2.vs.Gene.Expression <- Pseudo.lineage.2.vs.Gene.Expression[, percent_zero < 0.9]



# Apply the slope_corr function to Column A with all other columns
Pseudo.lineage.2.vs.Gene.Expression.Cor.df <- apply(Pseudo.lineage.2.vs.Gene.Expression[, -1], 2, function(x) slope_corr(Pseudo.lineage.2.vs.Gene.Expression$Lineage2, x))

# Convert results to dataframe
Pseudo.lineage.2.vs.Gene.Expression.Cor.df <- t(data.frame(Pseudo.lineage.2.vs.Gene.Expression.Cor.df))

colnames(Pseudo.lineage.2.vs.Gene.Expression.Cor.df) <- c("Slope", "Correlation", "P_value")

Pseudo.lineage.2.vs.Gene.Expression.Cor.df <- as.data.frame(Pseudo.lineage.2.vs.Gene.Expression.Cor.df)
```

```{r}
Pseudo.lineage.3.vs.Gene.Expression <- Pseudo.lineage.3 %>%
  left_join(Time.course.No.Day.0.Norm.Counts, by = "Cell") %>%
  dplyr::select(-c(Cluster.1.Score, Time.point, RNA_snn_res.0.6)) %>%
  column_to_rownames("Cell")   

# compute percentage of 0 values in each column
percent_zero <- apply(Pseudo.lineage.3.vs.Gene.Expression == 0, 2, mean)


# subset columns where less than 75% of values are 0, using the %in% operator
Pseudo.lineage.3.vs.Gene.Expression <- Pseudo.lineage.3.vs.Gene.Expression[, percent_zero < 0.9]

# Apply the slope_corr function to Column A with all other columns
Pseudo.lineage.3.vs.Gene.Expression.Cor.df <- apply(Pseudo.lineage.3.vs.Gene.Expression[, -1], 2, function(x) slope_corr(Pseudo.lineage.3.vs.Gene.Expression$Lineage3, x))

# Convert results to dataframe
Pseudo.lineage.3.vs.Gene.Expression.Cor.df <- t(data.frame(Pseudo.lineage.3.vs.Gene.Expression.Cor.df))

colnames(Pseudo.lineage.3.vs.Gene.Expression.Cor.df) <- c("Slope", "Correlation", "P_value")

Pseudo.lineage.3.vs.Gene.Expression.Cor.df <- as.data.frame(Pseudo.lineage.3.vs.Gene.Expression.Cor.df)
```

# Figure 4N
```{r}
p1 <- ggplot(Pseudo.lineage.1.vs.Gene.Expression, aes(x=Lineage1, y=Cxcl10)) + 
  geom_point() + stat_smooth(method = "lm") +
  background_grid(major = 'y', minor = "none") +
  panel_border() +
  stat_cor(aes(label = ..r.label..),method = "spearman", label.x = -0.5, label.y = 5, size = 4)  +
  ylab("Cxcl10 Normalized Counts") + 
  xlab("Pseudotime") +
  theme_bw() + ggtitle("Lineage 1") + theme(plot.title = element_text(hjust = 0.5), axis.title.x = element_text(size = 8), axis.title.y = element_text(size = 8)) +
  geom_vline(xintercept=16,linetype=3)

p2 <- ggplot(Pseudo.lineage.2.vs.Gene.Expression, aes(x=Lineage2, y=Saa3)) + 
  geom_point() + stat_smooth(method = "lm") +
  background_grid(major = 'y', minor = "none") +
  panel_border() +
  stat_cor(aes(label = ..r.label..),method = "spearman", label.x = -0.5, label.y = 5, size = 4)  +
  ylab("Saa3 Normalized Counts") +
  xlab("Pseudotime") +
  theme_bw() + ggtitle("Lineage 2") + theme(plot.title = element_text(hjust = 0.5), axis.title.x = element_text(size = 8), axis.title.y = element_text(size = 8))

p3 <- ggplot(Pseudo.lineage.3.vs.Gene.Expression, aes(x=Lineage3, y=Col1a1)) + 
  geom_point() + stat_smooth(method = "lm") +
  background_grid(major = 'y', minor = "none") +
  panel_border() +
  stat_cor(aes(label = ..r.label..),method = "spearman", label.x = -0.5, label.y = 5, size = 4)  +
  ylab("Col1a1 Normalized Counts") + 
  xlab("Pseudotime") +
  theme_bw() + ggtitle("Lineage 3") + theme(plot.title = element_text(hjust = 0.5, size = 10), axis.title.x = element_text(size = 8), axis.title.y = element_text(size = 8))

Fig_4N_plot <- plot_grid(p1, p2, p3, nrow = 3)

if (!dir.exists("Figure_4")) {dir.create("Figure_4")}
ggsave(plot = Fig_4N_plot, "Figure_4N_Top_Gene_Correlations.pdf", width = 3.75, height = 4.4444444, units = "in", path = "Figure_4/")
```

```{r}
Pseudo.lineage.1.vs.Gene.Expression.Cor.df <- Pseudo.lineage.1.vs.Gene.Expression.Cor.df %>%
  mutate(p.adj = p.adjust(Pseudo.lineage.1.vs.Gene.Expression.Cor.df$P_value, method = "bonferroni"))

Pseudo.lineage.2.vs.Gene.Expression.Cor.df <- Pseudo.lineage.2.vs.Gene.Expression.Cor.df %>%
  mutate(p.adj = p.adjust(Pseudo.lineage.2.vs.Gene.Expression.Cor.df$P_value, method = "bonferroni"))

Pseudo.lineage.3.vs.Gene.Expression.Cor.df <- Pseudo.lineage.3.vs.Gene.Expression.Cor.df %>%
  mutate(p.adj = p.adjust(Pseudo.lineage.3.vs.Gene.Expression.Cor.df$P_value, method = "bonferroni"))  

Pseudo.lineage.1.vs.Gene.Expression.Cor.Sig.Genes <- Pseudo.lineage.1.vs.Gene.Expression.Cor.df %>%
  filter(p.adj < 0.05 & Correlation > 0.25 | p.adj < 0.05 & Correlation < (-0.25)) 

Pseudo.lineage.2.vs.Gene.Expression.Cor.Sig.Genes <- Pseudo.lineage.2.vs.Gene.Expression.Cor.df %>%
  filter(p.adj < 0.05 & Correlation > 0.25 | p.adj < 0.05 & Correlation < (-0.25))

Pseudo.lineage.3.vs.Gene.Expression.Cor.Sig.Genes <- Pseudo.lineage.3.vs.Gene.Expression.Cor.df %>%
  filter(p.adj < 0.05 & Correlation > 0.25 | p.adj < 0.05 & Correlation < (-0.25))


```

```{r}
#write.table(Pseudo.lineage.1.vs.Gene.Expression.Cor.Sig.Genes, "Pseudo.lineage.1.vs.Gene.Expression.Cor.Sig.Genes.txt", sep = "\t")
#write.table(Pseudo.lineage.2.vs.Gene.Expression.Cor.Sig.Genes, "Pseudo.lineage.2.vs.Gene.Expression.Cor.Sig.Genes.txt", sep = "\t")
#write.table(Pseudo.lineage.3.vs.Gene.Expression.Cor.Sig.Genes, "Pseudo.lineage.3.vs.Gene.Expression.Cor.Sig.Genes.txt", sep = "\t")
```

# Gene set enrichment analysis
Gene set enrichment analysis is perfromed with clusterProfiler. 
```{r}
library(clusterProfiler)
library(org.Mm.eg.db)
library(ReactomePA)
library(msigdbr)
```

A function for the analysis of reactome GSEA based on correlations. It needs an input dataframe with DE genes including a column named `correlation` with correlation coefficients and the gene names as rownames. The function can be run in mode "reactome", "KEGG" or "custom" depending on the desired analysis. In custom mode, the gene set must be provided. 
The enrichment results are output as a dataframe
```{r}
GSEA_corr <-function(in_df, mode="reactome", custom_gene_set=NA){
  correl_df<-in_df %>% 
    rownames_to_column("Gene") %>% 
    dplyr::select(Gene, Correlation)
  
  gen_ID_key<-AnnotationDbi::select(org.Mm.eg.db, columns = c("ENTREZID", "SYMBOL"), keys = correl_df$Gene, keytype = "SYMBOL")
  
  correl_vector<-correl_df %>% 
    left_join(gen_ID_key, by=c("Gene"="SYMBOL")) %>% 
    drop_na() %>% 
    dplyr::select(ENTREZID, Correlation) %>% 
    arrange(desc(Correlation)) %>% 
    deframe()
  
  set.seed(1234)
  
  if (mode=="reactome"){
    gse <- gsePathway(geneList = correl_vector, organism = 'mouse', minGSSize = 10, pvalueCutoff = 0.05, pAdjustMethod = "BH", verbose = FALSE, seed = T)}
  
  if (mode=="KEGG"){
    gse <- gseKEGG(geneList = correl_vector, organism = 'mmu', minGSSize = 10, pvalueCutoff = 0.05, pAdjustMethod = "BH", verbose = FALSE, seed = T)}
  
  if (mode=="custom"){
    gse <- GSEA(geneList = correl_vector, TERM2GENE = custom_gene_set, minGSSize = 1, maxGSSize = 5000, pvalueCutoff = 0.05, pAdjustMethod = "BH", verbose = FALSE, seed = T)}
  
  gse_df<-gse@result %>%
    filter(p.adjust < 0.05)
  
  return(gse_df)
}
```




Specific gene set annotations are selected for custom ORA analysis
H: hallmark gene sets
C1: positional gene sets
C2: curated gene sets
C3: motif gene sets
C4: computational gene sets
C5: GO gene sets
C6: oncogenic signatures
C7: immunologic signatures
```{r}
H_Geneset <- msigdbr(species = "Mus musculus", category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)
C1_Geneset <- msigdbr(species = "Mus musculus", category = "C1") %>% 
  dplyr::select(gs_name, entrez_gene)
C2_Geneset <- msigdbr(species = "Mus musculus", category = "C2") %>% 
  dplyr::select(gs_name, entrez_gene)
C3_Geneset <- msigdbr(species = "Mus musculus", category = "C3") %>% 
  dplyr::select(gs_name, entrez_gene)
C4_Geneset <- msigdbr(species = "Mus musculus", category = "C4") %>% 
  dplyr::select(gs_name, entrez_gene)
C5_Geneset <- msigdbr(species = "Mus musculus", category = "C5") %>% 
  dplyr::select(gs_name, entrez_gene)
C6_Geneset <- msigdbr(species = "Mus musculus", category = "C6") %>% 
  dplyr::select(gs_name, entrez_gene)
C7_Geneset <- msigdbr(species = "Mus musculus", category = "C7") %>% 
  dplyr::select(gs_name, entrez_gene)
```

```{r}
Lin_1_Reactome_GSEA<-GSEA_corr(in_df = Pseudo.lineage.1.vs.Gene.Expression.Cor.Sig.Genes, mode = "reactome", custom_gene_set = NA)
Lin_1_KEGG_GSEA<-GSEA_corr(in_df = Pseudo.lineage.1.vs.Gene.Expression.Cor.Sig.Genes, mode = "KEGG", custom_gene_set = NA)
Lin_1_H_GSEA<-GSEA_corr(in_df = Pseudo.lineage.1.vs.Gene.Expression.Cor.Sig.Genes, mode = "custom", custom_gene_set = H_Geneset)

Lin_1_GSEA <- rbind(Lin_1_Reactome_GSEA, Lin_1_KEGG_GSEA, Lin_1_H_GSEA) %>%
  mutate(Lineage = "Lineage 1") %>%
  filter(NES > 0)
```

```{r}
Lin_2_Reactome_GSEA<-GSEA_corr(in_df = Pseudo.lineage.2.vs.Gene.Expression.Cor.Sig.Genes, mode = "reactome", custom_gene_set = NA)
Lin_2_KEGG_GSEA<-GSEA_corr(in_df = Pseudo.lineage.2.vs.Gene.Expression.Cor.Sig.Genes, mode = "KEGG", custom_gene_set = NA)
Lin_2_H_GSEA<-GSEA_corr(in_df = Pseudo.lineage.2.vs.Gene.Expression.Cor.Sig.Genes, mode = "custom", custom_gene_set = H_Geneset)

Lin_2_GSEA <- rbind(Lin_2_Reactome_GSEA, Lin_2_KEGG_GSEA, Lin_2_H_GSEA) %>%
  mutate(Lineage = "Lineage 2") %>%
  filter(NES > 0)
```

```{r}
Lin_3_Reactome_GSEA<-GSEA_corr(in_df = Pseudo.lineage.3.vs.Gene.Expression.Cor.Sig.Genes, mode = "reactome", custom_gene_set = NA)
Lin_3_KEGG_GSEA<-GSEA_corr(in_df = Pseudo.lineage.3.vs.Gene.Expression.Cor.Sig.Genes, mode = "KEGG", custom_gene_set = NA)
Lin_3_H_GSEA<-GSEA_corr(in_df = Pseudo.lineage.3.vs.Gene.Expression.Cor.Sig.Genes, mode = "custom", custom_gene_set = H_Geneset)

Lin_3_GSEA <- rbind(Lin_3_Reactome_GSEA, Lin_3_KEGG_GSEA, Lin_3_H_GSEA) %>%
  mutate(Lineage = "Lineage 3") %>%
  filter(NES > 0)
```


```{r}
Lin_1_GSEA_Enriched <- Lin_1_GSEA %>%
  dplyr::filter(p.adjust < 0.05)

Lin_2_GSEA_Enriched <- Lin_2_GSEA %>%
  dplyr::filter(p.adjust < 0.05)

Lin_3_GSEA_Enriched <- Lin_3_GSEA %>%
  dplyr::filter(p.adjust < 0.05)
```

```{r}
#write.table(Lin_1_GSEA_Enriched, "Time.course.Lin_1_GSEA_Enriched.txt", sep = "\t")
#write.table(Lin_2_GSEA_Enriched, "Time.course.Lin_2_GSEA_Enriched.txt", sep = "\t")
#write.table(Lin_3_GSEA_Enriched, "Time.course.Lin_3_GSEA_Enriched.txt", sep = "\t")
```


```{r}
Top_Lin_1_GSEA <- Lin_1_GSEA %>%
  top_n(20, NES)

Top_Lin_2_GSEA <- Lin_2_GSEA %>%
  top_n(20, NES)

Top_Lin_3_GSEA <- Lin_3_GSEA %>%
  top_n(20, NES)

`%!in%` = Negate(`%in%`)

Top_Lin_1_GSEA_unique <- Lin_1_GSEA %>%
  filter(Description %!in% c(Lin_2_GSEA$Description, Lin_3_GSEA$Description)) %>%
  top_n(20, NES)

Top_Lin_2_GSEA_unique <- Lin_2_GSEA %>%
  filter(Description %!in% c(Lin_1_GSEA$Description, Lin_3_GSEA$Description)) %>%
  top_n(20, NES)

Top_Lin_3_GSEA_unique <- Lin_3_GSEA %>%
  filter(Description %!in% c(Lin_2_GSEA$Description, Lin_1_GSEA$Description)) %>%
  top_n(20, NES)

Selected_Pathways <- c("NOD-like receptor signaling pathway - Mus musculus (house mouse)", "Herpes simplex virus 1 infection - Mus musculus (house mouse)", "Epstein-Barr virus infection - Mus musculus (house mouse)", "HALLMARK_INTERFERON_GAMMA_RESPONSE", "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_COMPLEMENT", "HALLMARK_HYPOXIA", "HALLMARK_INFLAMMATORY_RESPONSE", "TNF signaling pathway - Mus musculus (house mouse)", "Signaling by PDGF", "Assembly of collagen fibrils and other multimeric structures", "ECM proteoglycans", "Collagen biosynthesis and modifying enzymes", "Integrin cell surface interactions")

Lineage.GSEA <- rbind(Lin_1_GSEA, Lin_2_GSEA, Lin_3_GSEA)
Lineage.GSEA <- Lineage.GSEA %>%
  filter(Description %in% Selected_Pathways)
```

# Figure 4O
```{r}
Fig_4O_plot <- ggplot(Lineage.GSEA, aes(x = Lineage, y = Description)) + 
               geom_point(aes(color = p.adjust, size = NES)) +
               theme_bw(base_size =6) +
        scale_colour_gradient(limits=c(0, 0.05), low="red", high = "white") +
        #scale_size(range = c(2,8)) +
        ylab(NULL) +
        ggtitle("") +
  theme(axis.text.x = element_text(angle = 45, hjust=1, size = 10), title = element_text(size = 10), axis.text.y = element_text(size = 12)) +
  scale_x_discrete(limits = c("Lineage 1", "Lineage 2", "Lineage 3")) +
  scale_y_discrete(limits = c("NOD-like receptor signaling pathway - Mus musculus (house mouse)", "Herpes simplex virus 1 infection - Mus musculus (house mouse)", "Epstein-Barr virus infection - Mus musculus (house mouse)", "HALLMARK_INTERFERON_GAMMA_RESPONSE", "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_COMPLEMENT", "HALLMARK_HYPOXIA", "HALLMARK_INFLAMMATORY_RESPONSE", "TNF signaling pathway - Mus musculus (house mouse)", "Signaling by PDGF", "Assembly of collagen fibrils and other multimeric structures", "ECM proteoglycans", "Collagen biosynthesis and modifying enzymes", "Integrin cell surface interactions"))

if (!dir.exists("Figure_4")) {dir.create("Figure_4")}
ggsave(plot = Fig_4O_plot, "Figure_4O_GSEA_Lineage_Pseudotime.pdf", width = 11, height = 4.4444444, units = "in", path = "Figure_4/")
```


# Figure S7 - time course with day 0
The data is analyzed including day 0. The generation of the RDS files with selected singlets for each sample are described above. 
```{r}
S1 <- readRDS("S1_singlets.rds")
S2 <- readRDS("S2_singlets.rds")
S3 <- readRDS("S3_singlets.rds")
S4 <- readRDS("S4_singlets.rds")
S5 <- readRDS("S5_singlets.rds")
S6 <- readRDS("S6_singlets.rds")
S7 <- readRDS("S7_singlets.rds")
S8 <- readRDS("S8_singlets.rds")
S9 <- readRDS("S9_singlets.rds")
S10 <- readRDS("S10_singlets.rds")
S11 <- readRDS("S11_singlets.rds")
S12 <- readRDS("S12_singlets.rds")
S13 <- readRDS("S13_singlets.rds")
S14 <- readRDS("S14_singlets.rds")
S15 <- readRDS("S15_singlets.rds")
```

```{r}
Time.course <- merge(S1, c(S2, S3, S4, S5, S6, S7, S8, S9, S10, S11, S12, S13, S14, S15))
```

```{r}
VlnPlot(Time.course, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(Time.course, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Time.course, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

Time.course <- NormalizeData(Time.course)
Time.course <- FindVariableFeatures(Time.course, selection.method = "vst", nfeatures = 3000)
top10 <- head(VariableFeatures(Time.course), 10)
plot3 <- VariableFeaturePlot(Time.course)
plot4 <- LabelPoints(plot = plot3, points = top10, repel = TRUE)

Time.course <- ScaleData(Time.course, features = rownames(Time.course))
Time.course <- RunPCA(Time.course, features = VariableFeatures(object = Time.course))
VizDimLoadings(Time.course, dims = 1:2, reduction = "pca")
DimPlot(Time.course, reduction = "pca")
```

Cell cycle is regressed out. 
```{r}
mouse.s.genes<-c("Mcm5" ,"Pcna", "Tyms", "Fen1", "Mcm2", "Mcm4", "Rrm1", "Ung", "Gins2", "Mcm6", "Cdca7", "Dtl", "Prim1", "Uhrf1", "Mlf1ip", "Hells", "Rfc2", "Rpa2", "Nasp", "Rad51ap1", "Gmnn", "Wdr76", "Slbp", "Ccne2", "Ubr7", "Pold3", "Msh2", "Atad2", "Rad51", "Rrm2", "Cdc45", "Cdc6", "Exo1", "Tipin", "Dscc1", "Blm", "Casp8ap2", "Usp1", "Clspn", "Pola1", "Chaf1b", "Brip1", "E2f8")

mouse.g2m.genes<-c("Hmgb2", "Cdk1", "Nusap1", "Ube2c", "Birc5", "Tpx2", "Top2a", "Ndc80", "Cks2", "Nuf2", "Cks1b", "Mki67", "Tmpo", "Cenpf", "Tacc3", "Fam64a", "Smc4", "Ccnb2", "Ckap2l", "Ckap2", "Aurkb", "Bub1", "Kif11", "Anp32e", "Tubb4b", "Gtse1", "Kif20b", "Hjurp", "Cdca3", "Hn1", "Cdc20", "Ttk", "Cdc25c", "Kif2c", "Rangap1", "Ncapd2", "Dlgap5", "Cdca2", "Cdca8", "Ect2", "Kif23", "Hmmr", "Aurka", "Psrc1", "Anln", "Lbr", "Ckap5", "Cenpe", "Ctcf", "Nek2", "G2e3", "Gas2l3", "Cbx5", "Cenpa")

Time.course <- CellCycleScoring(Time.course, s.features = mouse.s.genes, g2m.features = mouse.g2m.genes, set.ident = TRUE)
RidgePlot(Time.course, features = c("Pcna", "Top2a", "Mcm6", "Mki67"), ncol = 2)

Time.course <- ScaleData(Time.course, vars.to.regress = c("S.Score", "G2M.Score", "percent.mt"), features = rownames(Time.course))
```

Principal component analysis
```{r}
Time.course <- RunPCA(Time.course, features = VariableFeatures(object = Time.course),  npcs = 100, verbose=FALSE)
```

```{r}
ElbowPlot(Time.course, ndims = 100)
```

Clustering and umap
```{r}
# Determine percent of variation associated with each PC
  pct <- Time.course@reductions$pca@stdev / sum(Time.course@reductions$pca@stdev) * 100
  # Determine the difference between variation of PC and subsequent PC
  co2 <- sort(which((pct[1:length(pct)-1] - pct[2:length(pct)]) > 0.1),  decreasing = T)[1] + 1 # last point    where change of % of variation is more than 0.1%.

Time.course <- FindNeighbors(Time.course, dims = 1:7)
Time.course <- FindClusters(Time.course, resolution = 0.6)
Time.course <- RunUMAP(Time.course, dims = 1:7)
# saveRDS(Time.course, "Time.course.Clustered.RDS")
```


```{r}
Time.course <- readRDS("Time.course.Clustered.RDS")
```

```{r}
pA <- DimPlot(Time.course, reduction = "umap", label = T) + NoLegend()
pB <- DimPlot(Time.course, reduction = "umap", group.by = "Time.point")

Time.point.cluster.prop <-  Time.course@meta.data %>% 
  group_by(RNA_snn_res.0.6, Time.point) %>% 
  summarise(count=n()) %>% 
  group_by(RNA_snn_res.0.6) %>% 
  mutate(prop=count/sum(count)) %>% 
  ungroup()

pC <- ggplot(Time.point.cluster.prop, aes(RNA_snn_res.0.6, y=prop, fill=Time.point)) +
  geom_bar(stat="identity", position="stack") +
  theme_bw() + 
  theme(panel.border = element_rect(colour = "black", fill = NA, size= 0.5),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank()) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 1)) +
  xlab("Cluster") +
  ylab("Proportion") + 
  scale_fill_discrete(name = "Time Point")
```


# Figure S7A-C
```{r}
Fig_S7AtoC <- cowplot::plot_grid(pA, pB, pC, nrow = 1)

if (!dir.exists("Figure_S7")) {dir.create("Figure_S7")}
ggsave(plot = Fig_S7AtoC, "Figure_S7A_C_Time_course_with_day_0.pdf", width = 15, height = 4.4444444, units = "in", path = "Figure_S7/")
```

SlingShot is used to predict trajectories
The objects are saved as separate matrices for input in Slingshot
```{r}
dimred <- Time.course@reductions$umap@cell.embeddings
clustering <- Time.course$RNA_snn_res.0.6
counts <- as.matrix(Time.course@assays$RNA@counts[Time.course@assays$RNA@var.features, ])
```

The default Slingshot lineage identification is performed
```{r}
set.seed(1)
lineages <- getLineages(data = dimred,
                        clusterLabels = clustering,
                        start.clus = "6", end.clus = c("0","5","4")) #define where to start the trajectories

lineages
```


# Figure S7D
```{r}
# Plot the lineages
ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}

color_list <- ggplotColours(n=7)
seur_pal <- color_list
plot(dimred[, 1:2], col = seur_pal[clustering], cex = 0.5, pch = 16) +
lines(SlingshotDataSet(lineages), lwd = 3, col = "black")
```

```{r}
Fig_S7D_plot <- ggplotify::as.ggplot(function() plot(dimred[, 1:2], col = seur_pal[clustering], cex = 0.5, pch = 16) +
lines(SlingshotDataSet(lineages), lwd = 3, col = "black"))

if (!dir.exists("Figure_S7")) {dir.create("Figure_S7")}
ggsave(plot = Fig_S7D_plot, "Figure_S7D_Time_course_with_day_0_Slingshot_Trajectory.pdf", width = 5, height = 4.4444444, units = "in", path = "Figure_S7/")
```


```{r}
sds <- slingshot(Embeddings(Time.course, "umap"), clusterLabels = Time.course@meta.data$RNA_snn_res.0.6, 
                 start.clus = "6", end.clus = c("0","5","4"))
```


```{r}
cell_pal <- function(cell_vars, pal_fun,...) {
  if (is.numeric(cell_vars)) {
    pal <- pal_fun(100, ...)
    return(pal[cut(cell_vars, breaks = 100)])
  } else {
    categories <- sort(unique(cell_vars))
    pal <- setNames(pal_fun(length(categories), ...), categories)
    return(pal[cell_vars])
  }
}
```

```{r}
pseudot <- slingPseudotime(sds)
```

#Figure S7E
```{r}
nc <- 2
pseudot <- slingPseudotime(sds)
nms <- colnames(pseudot)
nr <- ceiling(length(nms)/nc)
pal <- viridis(100, end = 0.95)
par(mfrow = c(nr, nc))
for (i in nms) {
  colors <- pal[cut(pseudot[,i], breaks = 100)]
  plot(reducedDim(SlingshotDataSet(sds)), col = colors, pch = 16, cex = 0.5, main = i)
  lines(SlingshotDataSet(sds), lwd = 2, col = 'black', type = 'lineages')
}
```


```{r}
colors <- pal[cut(pseudot[,"Lineage1"], breaks = 100)]
pA <- ggplotify::as.ggplot(function() plot(reducedDim(SlingshotDataSet(sds)), col = colors, pch = 16, cex = 0.5, main = "Lineage1") +
lines(SlingshotDataSet(sds), lwd = 2, col = 'black', type = 'lineages'))
colors <- pal[cut(pseudot[,"Lineage2"], breaks = 100)]
pB <- ggplotify::as.ggplot(function() plot(reducedDim(SlingshotDataSet(sds)), col = colors, pch = 16, cex = 0.5, main = "Lineage2") +
lines(SlingshotDataSet(sds), lwd = 2, col = 'black', type = 'lineages'))
colors <- pal[cut(pseudot[,"Lineage3"], breaks = 100)]
pC <- ggplotify::as.ggplot(function() plot(reducedDim(SlingshotDataSet(sds)), col = colors, pch = 16, cex = 0.5, main = "Lineage3") +
lines(SlingshotDataSet(sds), lwd = 2, col = 'black', type = 'lineages'))
colors <- pal[cut(pseudot[,"Lineage4"], breaks = 100)]
pD <- ggplotify::as.ggplot(function() plot(reducedDim(SlingshotDataSet(sds)), col = colors, pch = 16, cex = 0.5, main = "Lineage4") +
lines(SlingshotDataSet(sds), lwd = 2, col = 'black', type = 'lineages'))


```

# Figure S7E
```{r}
Fig_S7E_plot <- cowplot::plot_grid(pA, pB, pC, pD, nrow = 2, ncol = 2)

if (!dir.exists("Figure_S7")) {dir.create("Figure_S7")}
ggsave(plot = Fig_S7E_plot, "Figure_S7E_Time_course_with_day_0_Lineages.pdf", width = 10*1.5, height = 4.4444444*1.5, units = "in", path = "Figure_S7/")
```

```{r}
Pseudo.lineage.time.course<- as.data.frame(pseudot)

Time.course.meta.data <- Time.course@meta.data %>%
  rownames_to_column("Cell")

Pseudo.lineage.1.time.course <- Pseudo.lineage.time.course %>%
  dplyr::select(Lineage1)%>%
  drop_na() %>%
  rownames_to_column("Cell") %>% 
  left_join(Time.course.meta.data, by = "Cell") %>%
  dplyr::select(c(Time.point, Lineage1, RNA_snn_res.0.6, Cell))

Pseudo.lineage.2.time.course <- Pseudo.lineage.time.course %>%
  dplyr::select(Lineage2)%>%
  drop_na() %>%
  rownames_to_column("Cell") %>% 
  left_join(Time.course.meta.data, by = "Cell") %>%
  dplyr::select(c(Time.point, Lineage2, RNA_snn_res.0.6, Cell)) 

Pseudo.lineage.3.time.course <- Pseudo.lineage.time.course %>%
  dplyr::select(Lineage3)%>%
  drop_na() %>%
  rownames_to_column("Cell") %>% 
  left_join(Time.course.meta.data, by = "Cell") %>%
  dplyr::select(c(Time.point, Lineage3, RNA_snn_res.0.6, Cell)) 

Pseudo.lineage.4.time.course <- Pseudo.lineage.time.course %>%
  dplyr::select(Lineage4)%>%
  drop_na() %>%
  rownames_to_column("Cell") %>% 
  left_join(Time.course.meta.data, by = "Cell") %>%
  dplyr::select(c(Time.point, Lineage4, RNA_snn_res.0.6, Cell)) 

p1 <- ggplot(Pseudo.lineage.1.time.course, aes(x=Lineage1, y=Time.point, color = Time.point)) +
  geom_jitter() + theme_bw() + xlab("Pseudotime") + ggtitle("Lineage 1") + ylab("") + theme(legend.title = element_text(size = 0)) + theme(plot.title = element_text(hjust = 0.5)) +
  geom_vline(xintercept=16,linetype=3) + NoLegend()

p2 <- ggplot(Pseudo.lineage.2.time.course, aes(x=Lineage2, y=Time.point, color = Time.point)) +
  geom_jitter() + theme_bw() + xlab("Pseudotime") + ggtitle("Lineage 2") + ylab("") + theme(legend.title = element_text(size = 0)) + theme(plot.title = element_text(hjust = 0.5)) + NoLegend()

p3 <- ggplot(Pseudo.lineage.3.time.course, aes(x=Lineage3, y=Time.point, color = Time.point)) +
  geom_jitter() + theme_bw() + xlab("Pseudotime") + ggtitle("Lineage 3") + ylab("") + theme(legend.title = element_text(size = 0)) + theme(plot.title = element_text(hjust = 0.5)) + NoLegend()

p4 <- ggplot(Pseudo.lineage.4.time.course, aes(x=Lineage4, y=Time.point, color = Time.point)) +
  geom_jitter() + theme_bw() + xlab("Pseudotime") + ggtitle("Lineage 4") + ylab("") + theme(legend.title = element_text(size = 0)) + theme(plot.title = element_text(hjust = 0.5)) + NoLegend()
```

#Figure S7F
```{r}
Fig_S7F_plot <- cowplot::plot_grid(p1, p2, p3, p4, nrow = 2, ncol = 2)

if (!dir.exists("Figure_S7")) {dir.create("Figure_S7")}
ggsave(plot = Fig_S7F_plot, "Figure_S7F_Time_course_with_day_0_Lineages_pseudotime_vs_time_point.pdf", width = 7.5, height = 4.4444444, units = "in", path = "Figure_S7/")
```

The gene set enrichment analysis is performed with SCPA
```{r}
hall_mark_pathways <- msigdbr("mouse", "H")
reactome_pathways <- msigdbr(species = "mouse", category = "C2", subcategory = "CP:REACTOME")
kegg_pathways <- msigdbr(species = "mouse", category = "C2", subcategory = "CP:KEGG")

pathways <- rbind(hall_mark_pathways, reactome_pathways, kegg_pathways) %>%
  format_pathways()
```

```{r}
scpa_out_Day0_vs_Day1 <- compare_seurat(Time.course,
                           group1 = "Time.point", 
                           group1_population = c("Day.0", "Day.1"),
                           pathways = pathways)
```

```{r}
scpa_out_Day1_enriched <- scpa_out_Day0_vs_Day1 %>%
  filter(FC < 0 & adjPval < 0.05) %>%
  dplyr::mutate(Time.point = "Day 1")

top_scpa_out_Day1_enriched <- scpa_out_Day1_enriched %>%
  top_n(20, (qval + abs(FC)))

scpa_out_Day0_enriched <- scpa_out_Day0_vs_Day1 %>%
  filter(FC > 0 & adjPval < 0.05)  %>%
  dplyr::mutate(Time.point = "Day 0")

top_scpa_out_Day0_enriched <- scpa_out_Day0_enriched %>%
  top_n(20, (qval + abs(FC)))

Selected_Pathways <- c("REACTOME_TRANSLATION", "HALLMARK_MYC_TARGETS_V1", "REACTOME_RRNA_PROCESSING", "HALLMARK_E2F_TARGETS", "REACTOME_MRNA_SPLICING")

scpa_out_Day0_vs_Day1 <- rbind(scpa_out_Day1_enriched, scpa_out_Day0_enriched)

Day.1.vs.Day.0.Pathways <- scpa_out_Day0_vs_Day1 %>%
  filter(Pathway %in% Selected_Pathways)

Day.1.vs.Day.0.Pathways <- Day.1.vs.Day.0.Pathways %>%
  arrange(Time.point, desc(FC)) %>%
  mutate(Pathway = gsub("_", " ", Pathway))
```

# Figure S7G
```{r}
Fig_7G_plot <- ggplot(Day.1.vs.Day.0.Pathways, aes(x = Time.point, y = Pathway)) + 
               geom_point(aes(color = qval, size = abs(FC))) +
               theme_bw(base_size =6) +
        scale_colour_gradient(limits=c(0, 10), low="white", high = "red") +
        #scale_size(range = c(2,8)) +
        ylab(NULL) +
        ggtitle("") +
  theme(axis.text.x = element_text(angle = 45, hjust=1, size = 10), title = element_text(size = 10), axis.text.y = element_text(size = 12)) +
  scale_x_discrete(limits = c("Day 0", "Day 1")) +
  scale_y_discrete(limits = c("REACTOME TRANSLATION", "HALLMARK MYC TARGETS V1", "REACTOME RRNA PROCESSING", "HALLMARK E2F TARGETS", "REACTOME MRNA SPLICING"))

if (!dir.exists("Figure_S7")) {dir.create("Figure_S7")}
ggsave(plot = Fig_7G_plot, "Figure_S7G_GSEA_Day0_vs_Day1.pdf", width = 7.5, height = 4.4444444, units = "in", path = "Figure_S7/")
```

# Figure S7H
```{r}
Time.course.No.Day.0<-readRDS("Time.course.No.Day.0.Clustered.Annotated.Zscore.RDS")

Fig_7H_plot <- DimPlot(Time.course.No.Day.0, reduction = "umap", group.by = "Replicate")

if (!dir.exists("Figure_S7")) {dir.create("Figure_S7")}
ggsave(plot = Fig_7H_plot, "Figure_S7H_GSEA_UMAP_Replicate.pdf", width = 7.5, height = 4.4444444, units = "in", path = "Figure_S7/")
```

# Figure S7I
```{r}
Replicate.cluster.prop <-  Time.course.No.Day.0@meta.data %>% 
  group_by(RNA_snn_res.0.6, Replicate) %>% 
  summarise(count=n()) %>% 
  group_by(RNA_snn_res.0.6) %>% 
  mutate(prop=count/sum(count)) %>% 
  ungroup()

Fig_7I_plot <- ggplot(Replicate.cluster.prop, aes(x=factor(RNA_snn_res.0.6, levels = c('0', '3', '1', '2', '5', '4', '6')), y=prop, fill=Replicate)) +
  geom_bar(stat="identity", position="stack") +
  theme_bw() + 
  theme(panel.border = element_rect(colour = "black", fill = NA, size= 0.5),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank()) + 
    scale_y_continuous(expand = c(0, 0), limits = c(0, 1)) +
  xlab("Cluster") +
  ylab("Proportion") + 
  scale_fill_discrete(name = "Replicate")

if (!dir.exists("Figure_S7")) {dir.create("Figure_S7")}
ggsave(plot = Fig_7I_plot, "Figure_S7I_Replicate_Cluster_Prop.pdf", width = 3.75, height = 4.4444444, units = "in", path = "Figure_S7/")
```

# Figure S7J
```{r}
p1 <- VlnPlot(Time.course.No.Day.0, features = "Cluster 2 Co-culture Z-Score", pt.size = 0) + NoLegend() + labs(title = "Co-culture Crabp1+ CAF Z-score") & geom_hline(yintercept = 0) 

p4 <- FeaturePlot(Time.course.No.Day.0, features = "Cluster 2 Co-culture Z-Score", cols = c("lightblue", "red"))
```

```{r}
Fig_7J_plot <- plot_grid(p1, p4, nrow = 2)

if (!dir.exists("Figure_S7")) {dir.create("Figure_S7")}
ggsave(plot = Fig_7J_plot, "Figure_S7J_Coculture_Z_Scores.pdf", width = 3.75, height = 4.4444444, units = "in", path = "Figure_S7/")
```
