---
title: "FASTQ_to_RUC_18Maj2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Synopsis
The following code describes the processing of fastq files to oraw UMI count for the  single cell RNAseq data from 6 human PDAC samples where FAP+ mesenchymal cells were FACS-sorted for single cell sequencing. This code was used to produce figures 1, 2 and S1 to S5 of Cumming et al, 2025. 

The input data are the raw fastq files. The output is the raw umi count generated by the Cell Ranger pipeline. The raw fastq files are available via the Swedish National Data service (SND) upon request at https://doi.org/10.5878/0ehq-1434. The output of the present code (6 UMI count directories with count data in MEX format) are available openly on SND.

# Cell ranger pipeline
Cell Ranger performed. Samples 1-4 with default chemistry.
```{bash}
cellranger count --id=S1_counts --transcriptome=/data/henlab/software/cellranger_ref/refdata-gex-GRCh38-2020-A --fastqs=/corgi/joshuac/FAP_CAF_6S_10x/ --sample=S1 --expect-cells=8000 --localcores=10 --localmem=64
```
 
Sample 5-6 with multiomics chemistry because wrong beads where used. 
```{bash}
cellranger count --id=S5_counts --transcriptome=/data/henlab/software/cellranger_ref/refdata-gex-GRCh38-2020-A --fastqs=/corgi/joshuac/FAP_CAF_6S_10x/ --sample=S5 --chemistry=ARC-v1 --expect-cells=8000 --localcores=10 --localmem=64
```

The gene metadata, barcodes and expression matrices are retrieved from the cell ranger output. 
```{bash}
zcat S1_counts/outs/filtered_feature_bc_matrix/features.tsv.gz | cut -f1-2 > features.tsv

zcat S1_counts/outs/filtered_feature_bc_matrix/barcodes.tsv.gz | sed 's/-1//g' | awk '{print "S1_"$1}' > S1_barcodes

zcat S1_counts/outs/filtered_feature_bc_matrix/matrix.mtx.gz | awk 'NR>3' > S1_matrix.mtx
```

The gene metadata is retrieved 
```{r}
Gene_md<-read.table("features.tsv") %>% 
  rename(EnsID=V1, Gene = V2) %>% 
  mutate(Gene_nr = 1:nrow(.))
```

We check the gene name duplicates. 
```{r}
Gene_md %>% 
  group_by(Gene) %>% 
  filter(n()>1)
```

The gene with highest EnsID of the pair is renamed like Gene.1
```{r}
To_rename<-Gene_md %>% 
  group_by(Gene) %>% 
  filter(n()>1) %>% 
  arrange(Gene, desc(EnsID)) %>% 
  slice_head(n = 1) %>% 
  `$`(EnsID)

Gene_md<-Gene_md %>% 
  mutate(Gene=ifelse(EnsID %in% To_rename, paste(Gene, "1", sep = "."), Gene))
```

The barcode md are retrieved
```{r}
retrieve_bcs<-function(in_bc_file){
  bcs<-read.table(in_bc_file) %>% 
    mutate(Cell_nr=1:nrow(.)) %>% 
    rename(Cell = V1)
  
  return(bcs)
}

S1_bcs<-retrieve_bcs("S1_barcodes")
S2_bcs<-retrieve_bcs("S2_barcodes")
S3_bcs<-retrieve_bcs("S3_barcodes")
S4_bcs<-retrieve_bcs("S4_barcodes")
S5_bcs<-retrieve_bcs("S5_barcodes")
S6_bcs<-retrieve_bcs("S6_barcodes")
```


The count matrices are retrieved and the cell and gene info are integrated
```{r}
make_sample_matrix <- function(in_matrix_file, in_bcs, in_gene_md) {
  mtx <- read.table(in_matrix_file) %>%
    rename(Gene_nr = V1,
           Cell_nr = V2,
           RUC = V3)
  
  mtx <-
    sparseMatrix(
      i = mtx$Gene,
      j = mtx$Cell,
      x = mtx$RUC,
      dims = c(nrow(in_gene_md), nrow(in_bcs))
    )
  
  rownames(mtx) <- in_gene_md$Gene
  colnames(mtx) <- in_bcs$Cell
  
  return(mtx)
}


S1_mtx<-make_sample_matrix(in_matrix_file = "S1_matrix.mtx", in_bcs = S1_bcs, Gene_md)
S2_mtx<-make_sample_matrix(in_matrix_file = "S2_matrix.mtx", in_bcs = S2_bcs, Gene_md)
S3_mtx<-make_sample_matrix(in_matrix_file = "S3_matrix.mtx", in_bcs = S3_bcs, Gene_md)
S4_mtx<-make_sample_matrix(in_matrix_file = "S4_matrix.mtx", in_bcs = S4_bcs, Gene_md)
S5_mtx<-make_sample_matrix(in_matrix_file = "S5_matrix.mtx", in_bcs = S5_bcs, Gene_md)
S6_mtx<-make_sample_matrix(in_matrix_file = "S6_matrix.mtx", in_bcs = S6_bcs, Gene_md)
```

The matrices are merged to a single count matrix
```{r}
FapCAF_6S_mtx<-cbind(S1_mtx, S2_mtx, S3_mtx, S4_mtx, S5_mtx, S6_mtx)

FapCAF_6S_SO<-CreateSeuratObject(FapCAF_6S_mtx)
```





